<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>sPLS-DA and MINT models for cell type prediction and gene selection • BIRSBIO2020.scSpatial.sPLSDAMINT</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="sPLS-DA and MINT models for cell type prediction and gene selection">
<meta property="og:description" content="BIRSBIO2020.scSpatial.sPLSDAMINT">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">BIRSBIO2020.scSpatial.sPLSDAMINT</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/BIRSBIO2020.scSpatial.sPLS-DA_MINT.html">sPLS-DA and MINT models for cell type prediction and gene selection</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/fuerzhou/BIRSBIO2020.scSpatial.sPLSDA_MINT">
    <span class="fas fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>sPLS-DA and MINT models for cell type prediction and gene selection</h1>
                        <h4 class="author">Yuzhou Feng</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/fuerzhou/BIRSBIO2020.scSpatial.sPLSDA_MINT/blob/master/vignettes/BIRSBIO2020.scSpatial.sPLS-DA_MINT.Rmd"><code>vignettes/BIRSBIO2020.scSpatial.sPLS-DA_MINT.Rmd</code></a></small>
      <div class="hidden name"><code>BIRSBIO2020.scSpatial.sPLS-DA_MINT.Rmd</code></div>

    </div>

    
    

<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<div id="background" class="section level2">
<h2 class="hasAnchor">
<a href="#background" class="anchor"></a>Background</h2>
<p>Tasic et al. (2016) used scRNA sequencing to explore the extent of cell types of mouse primal visual cortex. In the study, 1723 high-quality cells with thousands of genes in primal visual cortex in male adult mice were sequenced on a single cell resolution for cell classification. This study identified 49 cell types, including 23 GABAergic, 19 glutamatergic and 7 non-neural types in the 1723 cells.</p>
<p>The cell classification result from Tasic et al. (2016) was used by Zhu et al. (2018) to design a study to distinguish the difference between intrinsic and extrinsic effect on gene expression. Intrinsic effect means the regulatory gene network, while extrinsic means the cellular microenvironment. The study was conducted by combining scRNA sequencing data and smFISH data (single-molecule fluorescence in situ hybridization). The former one has high molecular resolution of transcriptomics (thousands of genes) without spatial information, while the latter keeps the spatial information but loses the high resolution (only a few hundred genes).</p>
<p>Zhu et al. (2018) mapped the sRNA sequencing data to the seqFISH data to enhance the molecular resolution of the cells using SVM (support vector machine). The model was trained to identify the major cell type difference by training on scRNA data of 8 groups, GABAergic and Glutamatergic are the major neuron types, and other non-neuronal types, including Astrocytes, Mndothelial cells, microcytes, and three types of Oligocytes. The selected features (genes) were the top 43 differentially expressed (DE) genes in the identified 113 genes. The classification result was FURTHER validated by different evidence like cell type specific staining and previously reported marker genes.</p>
<p>Now, given a dataset of gene expression levels of 1723 cells and 113 genes and the cell types (from Tasic et al. 2016), our task is to predict the cells provided by Zhu et al. (2018), 1597 cells with 113 genes.</p>
<p><strong>Outline of analysis:</strong></p>
<p>We will adopt a semi-supervised approach to classify the cells of seqFISH data.</p>
<ol style="list-style-type: decimal">
<li>Train an sPLS-DA model on scRNA dataset.</li>
<li>Select genes by limiting the number of genes of “keepX” argument of each component during hyperparameter tuning.</li>
<li>Predict seqFISH data using the trained sPLS-DA model.</li>
<li>The predictions with high probabilities by the sPLS-DA model will be combined to the original training data for further training.</li>
<li>Train a MINT model on the combined training dataset and predict the rest cells of seqFISH data.</li>
<li>Identify the marker genes by examining the loading factors.</li>
</ol>
<p>This semi-supervised method can borrow the information from the seqFISH data and make the model customised. The combination of the data will be used to train a MINT.sPLS-DA model. MINT (Multivariate INTegrative method) (Rohart et al. 2017) is robust for integrating data from different sources regardless of the batch effect. The top discriminative genes identified from the model will be validated using previous literature as evidence. The determination of the minimal number of genes will be done by restricting the values of keepX and the performance will be monitored by balanced error rate (BER). BER is the average of the proportion of wrong classifications in each class. The lower the BER, the more accurate the model is.</p>
</div>
<div id="data" class="section level2">
<h2 class="hasAnchor">
<a href="#data" class="anchor"></a>Data</h2>
<p>First we import the data, including the training data (<code>scRNA</code>), the training labels (<code>scRNA_label</code>), the test data (<code>seqFISH</code>) and the predicted test labels by Zhu’s paper (<code>seqFISH_label</code>).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">##### import the data #####</span>
<span class="va">scRNA1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="st">"data/tasic_training_b2.txt"</span>,header <span class="op">=</span> <span class="cn">FALSE</span>, sep <span class="op">=</span><span class="st">'\t'</span><span class="op">)</span> 
<span class="va">scRNA</span> <span class="op">&lt;-</span> <span class="va">scRNA1</span><span class="op">[</span>,<span class="op">-</span><span class="fl">1</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">scRNA</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">scRNA1</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>
<span class="va">scRNA</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">scRNA</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">scRNA</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"scRNA"</span>,<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">scRNA</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>,<span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">scRNA</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span>
  
<span class="va">seqFISH1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="st">"data/seqfish_cortex_b2_testing.txt"</span>,header <span class="op">=</span> <span class="cn">FALSE</span>, sep <span class="op">=</span><span class="st">'\t'</span><span class="op">)</span> 
<span class="va">seqFISH</span> <span class="op">&lt;-</span> <span class="va">seqFISH1</span><span class="op">[</span>,<span class="op">-</span><span class="fl">1</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">seqFISH</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">seqFISH1</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>
<span class="va">seqFISH</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">seqFISH</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">seqFISH</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"seqFISH"</span>,<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">seqFISH</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>,<span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">seqFISH</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span>

<span class="va">scRNA_label</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="st">"data/tasic_labels.tsv"</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>
<span class="va">seqFISH_label</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="st">"data/seqfish_labels.tsv"</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span></code></pre></div>
<p><code>scRNA</code> is from Taisc et el. 2016. It contains the gene expression level of 1723 cells with 113 genes.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">scRNA</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 1723  113</code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Let's take a look at the first 6 cells.</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">scRNA</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>##         abca15 abca9 acta2 adcy4 aldh3b2 amigo2
## scRNA_1     11    22    15    12      27     23
## scRNA_2     42    46    47    45      49     43
## scRNA_3     17    22    15    12      27    101
## scRNA_4     42    46    42    45      49     43
## scRNA_5     35    39    34    38      42     72
## scRNA_6     35    39    42    38      42     43</code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">n_genes</span> <span class="op">&lt;-</span> <span class="fl">113</span>
<span class="va">n_scRNA</span> <span class="op">&lt;-</span> <span class="fl">1723</span></code></pre></div>
<p><code>seqFISH</code> is from Zhu et el. 2018. It contains the gene expression level of 1597 cells with the 113 genes shared with <code>scRNA</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">seqFISH</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 1597  113</code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Let's take a look at the first 6 cells.</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">seqFISH</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>##           abca15 abca9 acta2 adcy4 aldh3b2 amigo2
## seqFISH_1     68    41    25    39     101     93
## seqFISH_2     49    42    23    54      47     64
## seqFISH_3     50    38    16    37      41     93
## seqFISH_4     39    36    21    18      52     93
## seqFISH_5     31    47    29    37     101     93
## seqFISH_6     59    18    36    26      50     93</code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">n_seqFISH</span> <span class="op">&lt;-</span> <span class="fl">1597</span></code></pre></div>
<p><code>scRNA_label</code> has the class of the cells corresponding to <code>scRNA</code> data. The phenotyping was done by Tasic et al. and the labels are treated as ground truth in our analysis and are used for training models.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">scRNA_label</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 1723</code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Let's take a look at the number of cells in each type of the scRNA data</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">scRNA_label</span><span class="op">)</span></code></pre></div>
<pre><code>## scRNA_label
##            Astrocyte     Endothelial Cell    GABA-ergic Neuron 
##                   43                   29                  761 
## Glutamatergic Neuron            Microglia    Oligodendrocyte.1 
##                  812                   22                   19 
##    Oligodendrocyte.2    Oligodendrocyte.3 
##                    6                   31</code></pre>
<p>There are 8 classes in <code>scRNA_label</code>. The numbers of cells in each type are not balanced.</p>
<p><code>seqFISH_label</code> has the classes predicted for <code>seqFISH</code> data by Zhu et al. The predicitons have not been benchmarked, thus <code>seqFISH_label</code> will not be used in the prediction. It will be used to compare my prediction with Zhu’s prediction.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">seqFISH_label</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 1597</code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># the number of cells in each cell type of the seqFISH data predicted by Zhu's paper</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">seqFISH_label</span><span class="op">)</span></code></pre></div>
<pre><code>## seqFISH_label
##            Astrocyte     Endothelial Cell    GABA-ergic Neuron 
##                   97                   11                  556 
## Glutamatergic Neuron            Microglia    Oligodendrocyte.1 
##                  859                   22                    8 
##    Oligodendrocyte.2    Oligodendrocyte.3 
##                   21                   23</code></pre>
</div>
</div>
<div id="data-preprocessing" class="section level1">
<h1 class="hasAnchor">
<a href="#data-preprocessing" class="anchor"></a>Data preprocessing</h1>
<div id="quantile-normalisation" class="section level2">
<h2 class="hasAnchor">
<a href="#quantile-normalisation" class="anchor"></a>Quantile normalisation</h2>
<p>First, we check the gene expression distribution of each gene in <code>scRNA</code> and <code>seqFISH</code>.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">scRNA</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">scRNA</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span>, main <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">scRNA</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p><img src="hist_sc1.png"><img src="hist_sc2.png"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">seqFISH</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">seqFISH</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span>, main <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">seqFISH</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p><img src="hist_fish1.png"><img src="hist_fish2.png"></p>
<p>The histograms show that the features (genes) are of the same scale and the distributions are almost normal. It means that there’s no need for further feature normalisation.</p>
<p>Now we check the gene expression distribution of the 40 random samples in each dataset.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span>
<span class="va">random_scRNA</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">n_scRNA</span>, <span class="fl">40</span>, replace<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html">boxplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">scRNA</span><span class="op">)</span><span class="op">[</span>,<span class="va">random_scRNA</span><span class="op">]</span>, names <span class="op">=</span> <span class="va">random_scRNA</span>, 
        main <span class="op">=</span> <span class="st">"gene expression boxplot of 40 random scRNA samples"</span>, 
        xlab <span class="op">=</span> <span class="st">"sample ID (scRNA)"</span>,cex.main <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/unnamed-chunk-9-1.png" width="768"></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">random_seqFISH</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">n_seqFISH</span>, <span class="fl">40</span>, replace<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html">boxplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">seqFISH</span><span class="op">)</span><span class="op">[</span>,<span class="va">random_seqFISH</span><span class="op">]</span>, names <span class="op">=</span> <span class="va">random_seqFISH</span>, 
        main <span class="op">=</span> <span class="st">"gene expression boxplot of 40 random seqFISH samples"</span>, 
        xlab <span class="op">=</span> <span class="st">"sample ID (seqFISH)"</span>,cex.main <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/unnamed-chunk-9-2.png" width="768"></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Figure 1. Gene expression distribution of each sample in scRNA data and seqFISH data.</p>
<p>The boxplots show that the the distributions of scRNA data have some sample-wise variation, while seqFISH samples have more consistent distributions.</p>
<p>Some other analyses in the hackalthon did quantile normalisation for paired genes to try to eliminate the differences across datasets. However, we proposed to do <strong>quantile normalisation</strong> on <strong>samples</strong> because although different samples have different differentially expressed (DE) genes, the majority of the genes should express at the same level. That means whichever genes are DE, the mean and quantile of each sample should be of the similar level.</p>
<p>Therefore, We unify the quantile normalisation of samples between the datasets to maximumly eliminate the differences between the datasets. The normalisation is done by <code>normalize.quantiles</code> function from <code>preprocessCore</code> package.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## QUANTILE normalisation on all the samples</span>
<span class="va">all_data</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">scRNA</span>,<span class="va">seqFISH</span><span class="op">)</span>
<span class="va">qt_all_data</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu">normalize.quantiles</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">all_data</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">qt_all_data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">scRNA</span><span class="op">)</span>
<span class="co"># get the quantile normalised scRNA and seqFISH data</span>
<span class="va">qt_scRNA</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">qt_all_data</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">scRNA</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,<span class="op">]</span><span class="op">)</span>
<span class="va">qt_seqFISH</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">qt_all_data</span><span class="op">[</span><span class="op">(</span><span class="fl">1</span><span class="op">+</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">scRNA</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">scRNA</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">+</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">seqFISH</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span></code></pre></div>
<p>Then check 40 random samples from each dataset again.</p>
<p><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/unnamed-chunk-11-1.png" width="768"><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/unnamed-chunk-11-2.png" width="768"></p>
<p>Figure 2. Gene expression distribution of 40 random samples in scRNA data and seqFISH data after quantile normalisation.</p>
<p><code>qt_scRNA</code> is quantile normalised scRNA data and <code>qt_seqFISH</code> is quantile normalised seqFISH data.</p>
</div>
<div id="oversampling" class="section level2">
<h2 class="hasAnchor">
<a href="#oversampling" class="anchor"></a>Oversampling</h2>
<p>Let’s check the number of samples in each class.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">scRNA_label</span><span class="op">)</span></code></pre></div>
<pre><code>## scRNA_label
##            Astrocyte     Endothelial Cell    GABA-ergic Neuron 
##                   43                   29                  761 
## Glutamatergic Neuron            Microglia    Oligodendrocyte.1 
##                  812                   22                   19 
##    Oligodendrocyte.2    Oligodendrocyte.3 
##                    6                   31</code></pre>
<p>The cells of each class are heavily imbalanced in the training data. There is not enough training data for the minor group and the major groups will dominate the classification. Then the model will not have enough power to classify minor groups. To solve the problem, I used two strategies. First, I combined the three subgroups of Oligodendroctyes into one group. This is because in the Tasic paper, some cells were not identified as “core” cells for any Oligo but identified as “intermediate” cells and these three types were identified as close to each other. Also, Oligo.2 has too few samples. The second strategy is to oversample the minor groups.</p>
<p>First we combine the three Oligo subgroups.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scRNA_relabel</span> <span class="op">=</span> <span class="va">scRNA_label</span>
<span class="va">scRNA_relabel</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">scRNA_relabel</span> <span class="op">==</span> <span class="st">"Oligodendrocyte.1"</span><span class="op">)</span>,
                <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">scRNA_relabel</span> <span class="op">==</span> <span class="st">"Oligodendrocyte.2"</span><span class="op">)</span>,
                <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">scRNA_relabel</span> <span class="op">==</span> <span class="st">"Oligodendrocyte.3"</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="op">=</span> <span class="st">"Oligodendrocyte"</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">scRNA_relabel</span><span class="op">)</span></code></pre></div>
<pre><code>## scRNA_relabel
##            Astrocyte     Endothelial Cell    GABA-ergic Neuron 
##                   43                   29                  761 
## Glutamatergic Neuron            Microglia      Oligodendrocyte 
##                  812                   22                   56</code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">qt_scRNA_Astro</span> <span class="op">=</span> <span class="va">qt_scRNA</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">scRNA_label</span> <span class="op">==</span> <span class="st">"Astrocyte"</span><span class="op">)</span>,<span class="op">]</span>
<span class="va">qt_scRNA_Endo</span> <span class="op">=</span> <span class="va">qt_scRNA</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">scRNA_label</span> <span class="op">==</span> <span class="st">"Endothelial Cell"</span><span class="op">)</span>,<span class="op">]</span>
<span class="va">qt_scRNA_GABA</span> <span class="op">=</span> <span class="va">qt_scRNA</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">scRNA_label</span> <span class="op">==</span> <span class="st">"GABA-ergic Neuron"</span><span class="op">)</span>,<span class="op">]</span>
<span class="va">qt_scRNA_Glu</span> <span class="op">=</span> <span class="va">qt_scRNA</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">scRNA_label</span> <span class="op">==</span> <span class="st">"Glutamatergic Neuron"</span><span class="op">)</span>,<span class="op">]</span>
<span class="va">qt_scRNA_Micro</span> <span class="op">=</span> <span class="va">qt_scRNA</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">scRNA_label</span> <span class="op">==</span> <span class="st">"Microglia"</span><span class="op">)</span>,<span class="op">]</span>
<span class="va">qt_scRNA_O1</span> <span class="op">=</span> <span class="va">qt_scRNA</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">scRNA_label</span> <span class="op">==</span> <span class="st">"Oligodendrocyte.1"</span><span class="op">)</span>,<span class="op">]</span>
<span class="va">qt_scRNA_O2</span> <span class="op">=</span> <span class="va">qt_scRNA</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">scRNA_label</span> <span class="op">==</span> <span class="st">"Oligodendrocyte.2"</span><span class="op">)</span>,<span class="op">]</span>
<span class="va">qt_scRNA_O3</span> <span class="op">=</span> <span class="va">qt_scRNA</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">scRNA_label</span> <span class="op">==</span> <span class="st">"Oligodendrocyte.3"</span><span class="op">)</span>,<span class="op">]</span>
<span class="va">qt_scRNA_Oligo</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">qt_scRNA_O1</span>, <span class="va">qt_scRNA_O2</span>,<span class="va">qt_scRNA_O3</span><span class="op">)</span></code></pre></div>
<p>Now, we oversample the minor groups.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">com_qt_scRNA_label</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"Astrocyte"</span>, <span class="fl">43</span><span class="op">*</span><span class="fl">15</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"Endothelial Cell"</span>,<span class="fl">29</span><span class="op">*</span><span class="fl">25</span><span class="op">)</span>,
                       <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"GABA-ergic Neuron"</span>,<span class="fl">761</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"Glutamatergic Neuron"</span>, <span class="fl">812</span><span class="op">)</span>, 
                       <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"Microglia"</span>,<span class="fl">22</span><span class="op">*</span><span class="fl">30</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"Oligodendrocyte"</span>, <span class="fl">56</span><span class="op">*</span><span class="fl">12</span><span class="op">)</span><span class="op">)</span>
<span class="va">com_qt_scRNA_label</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">com_qt_scRNA_label</span><span class="op">)</span></code></pre></div>
<p>We end up with around 700 samples in each class. The total number of cells after oversampling is 4275.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">com_qt_scRNA_label</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 4275</code></pre>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">com_qt_scRNA_label</span><span class="op">)</span></code></pre></div>
<pre><code>## com_qt_scRNA_label
##            Astrocyte     Endothelial Cell    GABA-ergic Neuron 
##                  645                  725                  761 
## Glutamatergic Neuron            Microglia      Oligodendrocyte 
##                  812                  660                  672</code></pre>
<p><code>com_qt_scRNA_label</code> is the new scRNA label vector. Correspondingly, the <code>scRNA</code> data should also be resized. <code>com_qt_scRNA</code> is the oversampled scRNA data.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">com_qt_scRNA_Astro</span> <span class="op">=</span> <span class="va">qt_scRNA_Astro</span> <span class="op">%&gt;%</span> <span class="fu">slice</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu">n</span><span class="op">(</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">15</span><span class="op">)</span><span class="op">)</span>
<span class="va">com_qt_scRNA_Endo</span> <span class="op">=</span> <span class="va">qt_scRNA_Endo</span> <span class="op">%&gt;%</span> <span class="fu">slice</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu">n</span><span class="op">(</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">25</span><span class="op">)</span><span class="op">)</span>
<span class="va">com_qt_scRNA_GABA</span> <span class="op">=</span> <span class="va">qt_scRNA_GABA</span> 
<span class="va">com_qt_scRNA_Glu</span> <span class="op">=</span> <span class="va">qt_scRNA_Glu</span>
<span class="va">com_qt_scRNA_Micro</span> <span class="op">=</span> <span class="va">qt_scRNA_Micro</span> <span class="op">%&gt;%</span> <span class="fu">slice</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu">n</span><span class="op">(</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">30</span><span class="op">)</span><span class="op">)</span>
<span class="va">com_qt_scRNA_Oligo</span> <span class="op">=</span> <span class="va">qt_scRNA_Oligo</span> <span class="op">%&gt;%</span> <span class="fu">slice</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu">n</span><span class="op">(</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">12</span><span class="op">)</span><span class="op">)</span>
<span class="va">com_qt_scRNA</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">com_qt_scRNA_Astro</span>, <span class="va">com_qt_scRNA_Endo</span>, <span class="va">com_qt_scRNA_GABA</span>, <span class="va">com_qt_scRNA_Glu</span>,
                     <span class="va">com_qt_scRNA_Micro</span>, <span class="va">com_qt_scRNA_Oligo</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">com_qt_scRNA</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 4275  113</code></pre>
<p><code>com_qt_scRNA</code> has 113 cells.</p>
</div>
</div>
<div id="spls-da-model-training" class="section level1">
<h1 class="hasAnchor">
<a href="#spls-da-model-training" class="anchor"></a>sPLS-DA model training</h1>
<p>Sparse PLS discriminant analysis (sPLS-DA) is a supervised machine learning algorithm (Lê Cao et al. 2011). Its performance in public microarray and SNP data sets is similar to other algorithms and it’s very efficient. The function is available in <code>mixOmics</code> package.</p>
<p>Now let’s train the sPLS-DA model on preprocessed scRNA data. First we tune the number of components.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># The cross validation is stochastic so need to set seed for reproducibility</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">0610</span><span class="op">)</span>
<span class="co"># build the model</span>
<span class="va">com_qt_splsda</span> <span class="op">=</span> <span class="fu">splsda</span><span class="op">(</span><span class="va">com_qt_scRNA</span>, <span class="va">com_qt_scRNA_label</span>, ncomp <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>
<span class="co"># hyperparameter tuning</span>
<span class="va">com_qt_perf.splsda</span> <span class="op">&lt;-</span> <span class="fu">perf</span><span class="op">(</span><span class="va">com_qt_splsda</span>, validation <span class="op">=</span> <span class="st">"Mfold"</span>, 
                           folds <span class="op">=</span> <span class="fl">5</span>, nrepeat <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">com_qt_perf.splsda</span>, overlay <span class="op">=</span> <span class="st">'measure'</span><span class="op">)</span></code></pre></div>
<p><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/unnamed-chunk-19-1.png" width="576"></p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">n_comp</span> <span class="op">&lt;-</span> <span class="fl">6</span>  <span class="co"># 6 is the elbow in the error rate plot</span>
<span class="va">error_rate</span> <span class="op">&lt;-</span> <span class="va">com_qt_perf.splsda</span><span class="op">$</span><span class="va">error.rate</span><span class="op">$</span><span class="va">BER</span><span class="op">[</span><span class="va">n_comp</span>,<span class="st">"max.dist"</span><span class="op">]</span></code></pre></div>
<p>The error rate curve for tuning the number of components shows that the elbow is at comp 6, and at comp 6 <strong>max distance</strong> has the lowest error rate 0.0517155. Note that after oversampling, the balanced error rate (BER) is similar to overall error rate. We still use BER to evaluate the performance since the numbers of each class are still different.</p>
<p>Then, we tune the number of genes of each compnent. We use <strong>max distance</strong> as the distance method. (You can skip the next chunk and directly tune the model with fewer genes on each component using the code in “gene selection” section.)</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># tune keepX (the number of genes selected in each component)</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">0610</span><span class="op">)</span>
<span class="va">list.keepX</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">100</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span>
<span class="va">com_qt_splsda.tune</span> <span class="op">&lt;-</span> <span class="fu">tune.splsda</span><span class="op">(</span><span class="va">com_qt_scRNA</span>, <span class="va">com_qt_scRNA_label</span>, ncomp <span class="op">=</span> <span class="fl">6</span>, validation <span class="op">=</span> <span class="st">'Mfold'</span>, folds <span class="op">=</span> <span class="fl">5</span>, dist <span class="op">=</span> <span class="st">'max.dist'</span>,progressBar <span class="op">=</span> <span class="cn">F</span>  , test.keepX <span class="op">=</span> <span class="va">list.keepX</span>, nrepeat <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">com_qt_splsda.tune</span>, optimal <span class="op">=</span> <span class="cn">TRUE</span>, sd <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/unnamed-chunk-20-1.png" width="480"></p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># the number of genes selected in each component</span>
<span class="va">com_qt_splsda.tune</span><span class="op">$</span><span class="va">choice.keepX</span></code></pre></div>
<pre><code>## comp1 comp2 comp3 comp4 comp5 comp6 
##     5     1    65     1    75    10</code></pre>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">keepX</span> <span class="op">&lt;-</span> <span class="va">com_qt_splsda.tune</span><span class="op">$</span><span class="va">choice.keepX</span>

<span class="co"># the training error rate of the tuned model</span>
<span class="va">com_qt_splsda.tune</span><span class="op">$</span><span class="va">error.rate</span><span class="op">[</span><span class="fl">27</span>,<span class="va">n_comp</span><span class="op">]</span></code></pre></div>
<pre><code>## [1] 0.1037554</code></pre>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">error_rate</span> <span class="op">&lt;-</span> <span class="va">com_qt_splsda.tune</span><span class="op">$</span><span class="va">error.rate</span><span class="op">[</span><span class="fl">27</span>,<span class="va">n_comp</span><span class="op">]</span></code></pre></div>
<p>The number of selected genes in each component is 5, 1, 65, 1, 75, 10. The BER (balanced error rate, max distance) is 0.1037554.</p>
</div>
<div id="gene-selection" class="section level1">
<h1 class="hasAnchor">
<a href="#gene-selection" class="anchor"></a>Gene selection</h1>
<p>We can try using fewer genes for each component. Hence we restrict the number of genes of each component within 5.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># tune keepX</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">0610</span><span class="op">)</span>
<span class="va">list.keepX</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span>
<span class="va">select_genes.tune</span> <span class="op">&lt;-</span> <span class="fu">tune.splsda</span><span class="op">(</span><span class="va">com_qt_scRNA</span>, <span class="va">com_qt_scRNA_label</span>, ncomp <span class="op">=</span> <span class="fl">6</span>, validation <span class="op">=</span> <span class="st">'Mfold'</span>,folds <span class="op">=</span> <span class="fl">5</span>, dist <span class="op">=</span> <span class="st">'max.dist'</span>,progressBar <span class="op">=</span> <span class="cn">F</span>  , test.keepX <span class="op">=</span> <span class="va">list.keepX</span>, nrepeat <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>

<span class="co"># the number of genes selected in each component</span>
<span class="va">select_genes.tune</span><span class="op">$</span><span class="va">choice.keepX</span></code></pre></div>
<pre><code>## comp1 comp2 comp3 comp4 comp5 comp6 
##     5     1     5     5     5     5</code></pre>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">keepX</span> <span class="op">&lt;-</span> <span class="va">select_genes.tune</span><span class="op">$</span><span class="va">choice.keepX</span>

<span class="co"># the training error rate of the tuned model</span>
<span class="va">select_genes.tune</span><span class="op">$</span><span class="va">error.rate</span><span class="op">[</span><span class="fl">5</span>,<span class="va">n_comp</span><span class="op">]</span></code></pre></div>
<pre><code>## [1] 0.1014857</code></pre>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">error_rate</span> <span class="op">&lt;-</span> <span class="va">select_genes.tune</span><span class="op">$</span><span class="va">error.rate</span><span class="op">[</span><span class="fl">5</span>,<span class="va">n_comp</span><span class="op">]</span></code></pre></div>
<p>The tuned keepX is 5, 1, 5, 5, 5, 5. The BER (max distance) is 0.1014857, even lower than the model with more genes.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">select_genes.tune</span><span class="op">$</span><span class="va">error.rate.class</span><span class="op">[</span>,<span class="fl">6</span><span class="op">]</span></code></pre></div>
<pre><code>##            Astrocyte     Endothelial Cell    GABA-ergic Neuron 
##           0.00000000           0.07413793           0.18971748 
## Glutamatergic Neuron            Microglia      Oligodendrocyte 
##           0.11961207           0.00000000           0.22544643</code></pre>
<p>The error rate (training set) of all the classes are low. The training error rate of Astro and Micro are 0.0. This may be explained by the presence of evidently representative marker genes in these two types.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># this is the final sPLS-DA model (using the tuned hyperparameters)</span>
<span class="va">splsda</span> <span class="op">&lt;-</span> <span class="fu">mixOmics</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/mixOmics/man/splsda.html">splsda</a></span>
<span class="va">final_sPLSDA.train</span> <span class="op">&lt;-</span> <span class="fu">splsda</span><span class="op">(</span><span class="va">com_qt_scRNA</span>, <span class="va">com_qt_scRNA_label</span>, ncomp <span class="op">=</span> <span class="fl">6</span>, keepX <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5</span>,<span class="fl">1</span>,<span class="fl">5</span>,<span class="fl">5</span>,<span class="fl">5</span>,<span class="fl">5</span><span class="op">)</span><span class="op">)</span> 

<span class="co"># predict seqFISH</span>
<span class="va">final_sPLSDA.predict</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span><span class="op">(</span><span class="va">final_sPLSDA.train</span>, <span class="va">qt_seqFISH</span><span class="op">)</span>

<span class="co"># Take a look at the number of cells predicted in each class</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">final_sPLSDA.predict</span><span class="op">$</span><span class="va">class</span><span class="op">$</span><span class="va">max.dist</span><span class="op">[</span>, <span class="va">n_comp</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>## 
##            Astrocyte     Endothelial Cell    GABA-ergic Neuron 
##                  114                   21                  539 
## Glutamatergic Neuron            Microglia      Oligodendrocyte 
##                  797                   55                   71</code></pre>
</div>
<div id="select-the-predictions-with-high-probability" class="section level1">
<h1 class="hasAnchor">
<a href="#select-the-predictions-with-high-probability" class="anchor"></a>Select the predictions with high probability</h1>
<p>The predictions of the cell types with probability higher than 50% are considered as confident classifications. Hence we keep the cells that have prediction probability over 50%.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pred_prob</span> <span class="op">=</span> <span class="va">final_sPLSDA.predict</span><span class="op">[[</span><span class="st">"predict"</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>,,<span class="va">n_comp</span><span class="op">]</span>
<span class="va">low_prob</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu">rowMaxs</span><span class="op">(</span><span class="va">pred_prob</span><span class="op">)</span><span class="op">&lt;</span><span class="fl">0.5</span><span class="op">)</span>
<span class="co"># the seqFISH cells with high probabilities (predictions considered "ground truth")</span>
<span class="va">seqFISH_data_predicted</span> <span class="op">=</span> <span class="va">qt_seqFISH</span><span class="op">[</span><span class="op">-</span><span class="va">low_prob</span>,<span class="op">]</span> 
<span class="co"># the labels of the seqFISH cells with high probabilities</span>
<span class="va">seqFISH_label_predicted</span> <span class="op">=</span> <span class="va">final_sPLSDA.predict</span><span class="op">$</span><span class="va">class</span><span class="op">$</span><span class="va">max.dist</span><span class="op">[</span>, <span class="va">n_comp</span><span class="op">]</span><span class="op">[</span><span class="op">-</span><span class="va">low_prob</span><span class="op">]</span> 

<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">seqFISH_label_predicted</span><span class="op">)</span></code></pre></div>
<pre><code>## seqFISH_label_predicted
##            Astrocyte     Endothelial Cell    GABA-ergic Neuron 
##                   84                    7                  376 
## Glutamatergic Neuron            Microglia      Oligodendrocyte 
##                  663                   29                   35</code></pre>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">low_prob</span><span class="op">)</span> </code></pre></div>
<pre><code>## [1] 403</code></pre>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">n_low_prob</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">low_prob</span><span class="op">)</span></code></pre></div>
<p>403 cells are with low prediction prob (&lt; 0.5).</p>
<p>We compare the sPLSDA predictions with Zhu paper.</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># my prediction</span>
<span class="va">predictions</span> <span class="op">=</span> <span class="va">final_sPLSDA.predict</span><span class="op">$</span><span class="va">class</span><span class="op">$</span><span class="va">max.dist</span><span class="op">[</span>, <span class="va">n_comp</span><span class="op">]</span>

<span class="co"># Zhu's prediction</span>
<span class="va">seqFISH_relabel</span> <span class="op">=</span> <span class="va">seqFISH_label</span>
<span class="va">seqFISH_relabel</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">seqFISH_relabel</span> <span class="op">==</span> <span class="st">"Oligodendrocyte.1"</span><span class="op">)</span>, 
                  <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">seqFISH_relabel</span> <span class="op">==</span> <span class="st">"Oligodendrocyte.2"</span><span class="op">)</span>, 
                  <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">seqFISH_relabel</span> <span class="op">==</span> <span class="st">"Oligodendrocyte.3"</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="op">=</span> <span class="st">"Oligodendrocyte"</span>

<span class="co"># the percentage of cells that are the same with Zhu's prediction</span>
<span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">seqFISH_relabel</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">predictions</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">predictions</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 0.7069505</code></pre>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">converge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">seqFISH_relabel</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">predictions</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">predictions</span><span class="op">)</span></code></pre></div>
<p>71% of the final predictions of splsda converge with Zhu paper.</p>
<p>The above model used max distance when tuning the hyperparameters. We also tuned the model with other distances. The code for tuning is shown below.</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#centroids.tune &lt;- tune.splsda(com_qt_scRNA, com_qt_scRNA_label, ncomp = 6, validation = 'Mfold',</span>
<span class="co">#                              folds = 5, dist = 'centroids.dist',progressBar = F, </span>
<span class="co">#                              test.keepX = list.keepX, nrepeat = 8)</span>
<span class="co">#mahalanobis.tune &lt;- tune.splsda(com_qt_scRNA, com_qt_scRNA_label, ncomp = 6, validation = 'Mfold', </span>
<span class="co">#                                folds = 5, dist = 'mahalanobis.dist',progressBar = F  , </span>
<span class="co">#                                test.keepX = list.keepX, nrepeat = 8)</span>
<span class="co">#centroids.tune$choice.keepX</span>
<span class="co">#centroids.tune$error.rate[5,6]</span>
<span class="co">#mahalanobis.tune$choice.keepX</span>
<span class="co">#mahalanobis.tune$error.rate[5,6]</span></code></pre></div>
<p>You can use these two models to predict the seqFISH data simply by modifying the code from line 294. The results will show that the convergence with Zhu’s paper of the two models is the same to the model with max distance.</p>
<p>Now we calculate the total number of genes selected in the sPLS-DA model.</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># calculate the number of selected features</span>
<span class="va">features</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">n_comp</span><span class="op">)</span><span class="op">{</span>
  <span class="va">new_features</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="fu">plotLoadings</span><span class="op">(</span><span class="va">final_sPLSDA.train</span>, plot <span class="op">=</span> <span class="cn">F</span>,
                                       contrib <span class="op">=</span> <span class="st">"max"</span>, method <span class="op">=</span> <span class="st">"median"</span>,comp <span class="op">=</span> <span class="va">i</span><span class="op">)</span><span class="op">)</span>
  <span class="va">features</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">features</span>,<span class="va">new_features</span><span class="op">)</span>
<span class="op">}</span>
<span class="va">features</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">features</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">features</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 24</code></pre>
<p>With only 24 genes the model is able to accurately classify the training data and make the predictions converge with Zhu’s paper up to 0.7069505.</p>
<p>The following figures are the sample plots of the sPLS-DA model. The plots are also overlaid with the predicted cell types of seqFISH data.</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">##### sampleplots ####</span>
<span class="co"># extract prediction labels </span>
<span class="va">label.predict</span> <span class="op">&lt;-</span><span class="va">final_sPLSDA.predict</span><span class="op">$</span><span class="va">class</span><span class="op">$</span><span class="va">max.dist</span><span class="op">[</span>, <span class="st">'comp6'</span><span class="op">]</span>
<span class="va">label.predict</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">label.predict</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">com_qt_scRNA_label</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">label.predict</span><span class="op">)</span></code></pre></div>
<pre><code>##            Astrocyte     Endothelial Cell    GABA-ergic Neuron 
##                  114                   21                  539 
## Glutamatergic Neuron            Microglia      Oligodendrocyte 
##                  797                   55                   71</code></pre>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">col.predict</span> <span class="op">&lt;-</span> <span class="fu">color.mixo</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">label.predict</span><span class="op">)</span><span class="op">)</span>

<span class="co"># comp 1&amp;2 with legend</span>
<span class="fu">plotIndiv</span><span class="op">(</span><span class="va">final_sPLSDA.train</span>, ellipse <span class="op">=</span> <span class="cn">TRUE</span>,comp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>,legend <span class="op">=</span> <span class="cn">TRUE</span>, legend.title <span class="op">=</span> <span class="st">'Cell type'</span>,
          ind.names <span class="op">=</span> <span class="cn">FALSE</span>, style <span class="op">=</span> <span class="st">"graphics"</span>, pch <span class="op">=</span> <span class="fl">16</span>, ellipse.level <span class="op">=</span> <span class="fl">0.90</span>,
          X.label <span class="op">=</span> <span class="st">'sPLS-DA comp1'</span>, Y.label <span class="op">=</span> <span class="st">'sPLS-DA comp2'</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/points.html">points</a></span><span class="op">(</span><span class="va">final_sPLSDA.predict</span><span class="op">$</span><span class="va">variates</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, <span class="va">final_sPLSDA.predict</span><span class="op">$</span><span class="va">variates</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, pch <span class="op">=</span> <span class="fl">2</span>, 
       cex <span class="op">=</span> <span class="fl">0.5</span>, col <span class="op">=</span> <span class="va">col.predict</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span><span class="op">(</span><span class="st">"topleft"</span>,legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"scRNA training"</span>,<span class="st">"seqFISH prediction"</span><span class="op">)</span>, pch <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">16</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/unnamed-chunk-28-1.png" width="700"></p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># comp 1&amp;3 with legend</span>
<span class="fu">plotIndiv</span><span class="op">(</span><span class="va">final_sPLSDA.train</span>, ellipse <span class="op">=</span> <span class="cn">TRUE</span>,comp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">4</span><span class="op">)</span>,legend <span class="op">=</span> <span class="cn">TRUE</span>, legend.title <span class="op">=</span> <span class="st">'Cell type'</span>,
          ind.names <span class="op">=</span> <span class="cn">FALSE</span>, style <span class="op">=</span> <span class="st">"graphics"</span>, pch <span class="op">=</span> <span class="fl">16</span>, ellipse.level <span class="op">=</span> <span class="fl">0.90</span>, 
          X.label <span class="op">=</span> <span class="st">'sPLS-DA comp1'</span>, Y.label <span class="op">=</span> <span class="st">'sPLS-DA comp3'</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/points.html">points</a></span><span class="op">(</span><span class="va">final_sPLSDA.predict</span><span class="op">$</span><span class="va">variates</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, <span class="va">final_sPLSDA.predict</span><span class="op">$</span><span class="va">variates</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span>, pch <span class="op">=</span> <span class="fl">2</span>, 
       cex <span class="op">=</span> <span class="fl">0.5</span>, col <span class="op">=</span> <span class="va">col.predict</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span><span class="op">(</span><span class="st">"topleft"</span>,legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"scRNA training"</span>,<span class="st">"seqFISH prediction"</span><span class="op">)</span>, pch <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">16</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/unnamed-chunk-29-1.png" width="700"></p>
<p>The sample plots show that on the first 3 components, the sPLS-DA model can discriminate Astrocyte, Microglia and Endothelial cell types in scRNA data (circles) well. Overlaying seqFish data (triangels) shows that the predictions of cell types for seqFISH by sPLS-DA model are also in agreement with the cell types from scRNA data.</p>
<p>The codes for sample plots of other components can be adapted from the codes for the first three components.</p>
</div>
<div id="identification-of-selected-genes" class="section level1">
<h1 class="hasAnchor">
<a href="#identification-of-selected-genes" class="anchor"></a>Identification of selected genes</h1>
<p>Let’s plot the loading factors on each component. It shows how much the genes contribute to the class discrimination of the training data on each component.For example in the following plots, the contribution on Comp 1 shows that gene <strong>gja1</strong> is over expressed in Astrocyte class according to its median value compared to the other cell types.</p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## loading factors for the trained splsda model</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span>
<span class="fu">plotLoadings</span><span class="op">(</span><span class="va">final_sPLSDA.train</span>, comp <span class="op">=</span> <span class="fl">1</span>, method <span class="op">=</span> <span class="st">'median'</span>, contrib <span class="op">=</span> <span class="st">'max'</span>,cex.main <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> </code></pre></div>
<p><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/unnamed-chunk-30-1.png" width="768"></p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">plotLoadings</span><span class="op">(</span><span class="va">final_sPLSDA.train</span>, comp <span class="op">=</span> <span class="fl">2</span>, method <span class="op">=</span> <span class="st">'median'</span>, contrib <span class="op">=</span> <span class="st">'max'</span>,legend <span class="op">=</span> <span class="cn">FALSE</span>, title <span class="op">=</span> <span class="st">"comp2"</span>,cex.main <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>
<span class="fu">plotLoadings</span><span class="op">(</span><span class="va">final_sPLSDA.train</span>, comp <span class="op">=</span> <span class="fl">3</span>, method <span class="op">=</span> <span class="st">'median'</span>, contrib <span class="op">=</span> <span class="st">'max'</span>,legend <span class="op">=</span> <span class="cn">FALSE</span>, title <span class="op">=</span> <span class="st">"comp3"</span>,cex.main <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></code></pre></div>
<p><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/unnamed-chunk-30-2.png" width="768"></p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">plotLoadings</span><span class="op">(</span><span class="va">final_sPLSDA.train</span>, comp <span class="op">=</span> <span class="fl">4</span>, method <span class="op">=</span> <span class="st">'median'</span>, contrib <span class="op">=</span> <span class="st">'max'</span>,legend <span class="op">=</span> <span class="cn">FALSE</span>, title <span class="op">=</span> <span class="st">"comp4"</span>,cex.main <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>
<span class="fu">plotLoadings</span><span class="op">(</span><span class="va">final_sPLSDA.train</span>, comp <span class="op">=</span> <span class="fl">5</span>, method <span class="op">=</span> <span class="st">'median'</span>, contrib <span class="op">=</span> <span class="st">'max'</span>,legend <span class="op">=</span> <span class="cn">FALSE</span>, title <span class="op">=</span> <span class="st">"comp5"</span>,cex.main <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></code></pre></div>
<p><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/unnamed-chunk-30-3.png" width="768"></p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">plotLoadings</span><span class="op">(</span><span class="va">final_sPLSDA.train</span>, comp <span class="op">=</span> <span class="fl">6</span>, method <span class="op">=</span> <span class="st">'median'</span>, contrib <span class="op">=</span> <span class="st">'max'</span>,legend <span class="op">=</span> <span class="cn">FALSE</span>, title <span class="op">=</span> <span class="st">"comp6"</span>,cex.main <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/unnamed-chunk-30-4.png" width="768"></p>
<p>The following table lists the most indictive genes for each cell type. The table also shows which genes are selected on which sPLS-DA component..</p>
<p><img src="scRNA_markergene.png"></p>
<p>Regarding the list of genes selected, several were found to be associated with specific cell types: <strong>Gja1</strong> was reported to regulate <strong>astrocytic</strong> migration and proliferation (Homkajorn et al. 2010) and was identified as a marker gene of <strong>astrocytes</strong> in the original study of Zhu et al. (2018) ; <strong>Cldn5</strong> encodes a protein that forms tight junction strands, as a physical barrier to prevent solutes and water from passing through <strong>endothelial</strong> cell sheets (Barmeyer et al. 2017); <strong>Omg</strong> (oligodendrocyte myelin glycoprotein) is present in <strong>Oligodendrocytes</strong>; <strong>Laptm5</strong> was chosen in Bonham et al. 2019 to illustrate the expression profiles for the <strong>microglial</strong> gene module and <strong>Tbr1</strong> was found to be a common genetic determinant for the differentiation of early-born <strong>glutamatergic</strong> neocortical neurons (Hevner et al. 2001).</p>
<p>In general, the genes found with large loading factors by the sPLS-DA model are backed up with previous studies. The classifier has a biological meaning.</p>
</div>
<div id="prediction-of-cells-from-seqfish-with-low-probability-using-the-mint-model" class="section level1">
<h1 class="hasAnchor">
<a href="#prediction-of-cells-from-seqfish-with-low-probability-using-the-mint-model" class="anchor"></a>Prediction of cells from seqFISH with low probability using the MINT model</h1>
<p>The PCA sample plot shows that there is batch effect between training and test datasets. To account for the batch effect, we will use a semi-supervised approach to borrow the information from seqFISH data. MINT algorithm is able to consider the batch effect between datasets. The predictions with high probabilities by sPLS-DA model will be considered as accurate labels here. I will train a MINT model on the combination of the training data and the predictions of seqFISH with high probabilities and predict the seqFISH data that have low probabilities.</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># pca sample plot for seqFISH and scRNA</span>
<span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">qt_scRNA</span>, <span class="va">qt_seqFISH</span><span class="op">)</span>
<span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"scRNA"</span>,<span class="va">n_scRNA</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"seqFISH"</span>,<span class="va">n_seqFISH</span><span class="op">)</span><span class="op">)</span>
<span class="va">pca</span> <span class="op">&lt;-</span> <span class="fu">pca</span><span class="op">(</span><span class="va">X</span>, ncomp <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="fu">plotIndiv</span><span class="op">(</span><span class="va">pca</span>, group <span class="op">=</span> <span class="va">group</span>, legend <span class="op">=</span> <span class="cn">T</span>, ellipse <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></code></pre></div>
<p><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/pca-1.png" width="480"></p>
<p>We combine the “accurate” seqFISH data and scRNA data. The combined data still need oversampling and undersampling due to the imbalance of classes.</p>
<pre><code>## [1] 2917  113</code></pre>
<pre><code>## [1] 2917</code></pre>
<pre><code>## [1] 2917</code></pre>
<pre><code>## Y
##            Astrocyte     Endothelial Cell    GABA-ergic Neuron 
##                  127                   36                 1137 
## Glutamatergic Neuron            Microglia      Oligodendrocyte 
##                 1475                   51                   91</code></pre>
<p>Now we oversample the minor groups and undersample the major groups. The number of samples in each groups will be around 500.</p>
<pre><code>## [1] 2998  113</code></pre>
<pre><code>## Y
##            Astrocyte     Endothelial Cell    GABA-ergic Neuron 
##                  508                  576                  500 
## Glutamatergic Neuron            Microglia      Oligodendrocyte 
##                  500                  459                  455</code></pre>
<pre><code>## [1] 2998</code></pre>
<p>We train a MINT.sPLSDA model on the combined data.</p>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">0610</span><span class="op">)</span>
<span class="va">mint</span> <span class="op">=</span> <span class="fu">mint.splsda</span><span class="op">(</span><span class="va">X</span>, <span class="va">Y</span>, study <span class="op">=</span> <span class="va">study</span>, ncomp <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>
<span class="co"># tune the number of component</span>
<span class="va">perf.mint.splsda</span> <span class="op">&lt;-</span> <span class="fu">perf</span><span class="op">(</span><span class="va">mint</span>, validation <span class="op">=</span> <span class="st">"Mfold"</span>, folds <span class="op">=</span> <span class="fl">5</span>,progressBar <span class="op">=</span> <span class="cn">FALSE</span>, nrepeat <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>
<span class="va">n_comp</span> <span class="op">=</span> <span class="fl">6</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">perf.mint.splsda</span>, overlay <span class="op">=</span> <span class="st">'measure'</span><span class="op">)</span></code></pre></div>
<p><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/unnamed-chunk-33-1.png" width="480"></p>
<p>The error rate curve for tuning the number of components shows that 6 component is enough to reach a low error rate. Now we use the number of genes selected by each component. We restrict the number of each component no more than 5.</p>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># tune keepX</span>
<span class="va">list.keepX</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">0610</span><span class="op">)</span>
<span class="va">mint.splsda.tune</span> <span class="op">&lt;-</span> <span class="fu">tune.mint.splsda</span><span class="op">(</span><span class="va">X</span>, <span class="va">Y</span>, ncomp <span class="op">=</span> <span class="fl">6</span>, dist <span class="op">=</span> <span class="st">'max.dist'</span>, study <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">study</span><span class="op">)</span>, measure <span class="op">=</span> <span class="st">"BER"</span>, test.keepX <span class="op">=</span> <span class="va">list.keepX</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">mint.splsda.tune</span>, optimal <span class="op">=</span> <span class="cn">TRUE</span>, sd <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/unnamed-chunk-34-1.png" width="576"></p>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># the chosen keepX after tuning</span>
<span class="va">mint.splsda.tune</span><span class="op">$</span><span class="va">choice.keepX</span></code></pre></div>
<pre><code>## comp1 comp2 comp3 comp4 comp5 comp6 
##     5     1     1     2     5     3</code></pre>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># the </span>
<span class="va">mint.splsda.tune</span><span class="op">$</span><span class="va">error.rate</span><span class="op">[</span><span class="fl">5</span>,<span class="va">n_comp</span><span class="op">]</span></code></pre></div>
<pre><code>## [1] 0.1565505</code></pre>
<p>Here is the final model after tuning.</p>
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## train mint</span>
<span class="va">mint.train</span> <span class="op">&lt;-</span> <span class="fu">mint.splsda</span><span class="op">(</span><span class="va">X</span>, <span class="va">Y</span>, ncomp <span class="op">=</span> <span class="va">n_comp</span>, study <span class="op">=</span> <span class="va">study</span>, 
                              keepX<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5</span>,<span class="fl">1</span>,<span class="fl">4</span>,<span class="fl">2</span>,<span class="fl">5</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>We plot the loading factors on each component.</p>
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span>
<span class="fu">plotLoadings</span><span class="op">(</span><span class="va">mint.train</span>, comp <span class="op">=</span> <span class="fl">1</span>, method <span class="op">=</span> <span class="st">'median'</span>, contrib <span class="op">=</span> <span class="st">'max'</span>,title <span class="op">=</span> <span class="st">"contribution on comp1"</span><span class="op">)</span> </code></pre></div>
<p><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/unnamed-chunk-37-1.png" width="768"></p>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">plotLoadings</span><span class="op">(</span><span class="va">mint.train</span>, comp <span class="op">=</span> <span class="fl">2</span>, method <span class="op">=</span> <span class="st">'median'</span>, contrib <span class="op">=</span> <span class="st">'max'</span> ,legend <span class="op">=</span> <span class="cn">FALSE</span>, title <span class="op">=</span> <span class="st">"comp2"</span>,cex.main <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>
<span class="fu">plotLoadings</span><span class="op">(</span><span class="va">mint.train</span>, comp <span class="op">=</span> <span class="fl">3</span>, method <span class="op">=</span> <span class="st">'median'</span>, contrib <span class="op">=</span> <span class="st">'max'</span> ,legend <span class="op">=</span> <span class="cn">FALSE</span>, title <span class="op">=</span> <span class="st">"comp3"</span>,cex.main <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></code></pre></div>
<p><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/unnamed-chunk-37-2.png" width="768"></p>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">plotLoadings</span><span class="op">(</span><span class="va">mint.train</span>, comp <span class="op">=</span> <span class="fl">4</span>, method <span class="op">=</span> <span class="st">'median'</span>, contrib <span class="op">=</span> <span class="st">'max'</span> ,legend <span class="op">=</span> <span class="cn">FALSE</span>, title <span class="op">=</span> <span class="st">"comp4"</span>,cex.main <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>
<span class="fu">plotLoadings</span><span class="op">(</span><span class="va">mint.train</span>, comp <span class="op">=</span> <span class="fl">5</span>, method <span class="op">=</span> <span class="st">'median'</span>, contrib <span class="op">=</span> <span class="st">'max'</span> ,legend <span class="op">=</span> <span class="cn">FALSE</span>, title <span class="op">=</span> <span class="st">"comp5"</span>,cex.main <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></code></pre></div>
<p><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/unnamed-chunk-37-3.png" width="768"></p>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">plotLoadings</span><span class="op">(</span><span class="va">mint.train</span>, comp <span class="op">=</span> <span class="fl">6</span>, method <span class="op">=</span> <span class="st">'median'</span>, contrib <span class="op">=</span> <span class="st">'max'</span> ,legend <span class="op">=</span> <span class="cn">FALSE</span>, title <span class="op">=</span> <span class="st">"comp6"</span>,cex.main <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/unnamed-chunk-37-4.png" width="768"></p>
<p><img src="mint_markergene.png"> The most predictive genes significantly overlap with the genes identified by sPLS-DA model. It means that the gene expression has a larger effect on the prediction rather than the batch effect. It also means that the MINT classifier also has biological meaning. Note that for GABA, there are no significantly high loading-factor genes.</p>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">plotIndiv</span><span class="op">(</span><span class="va">mint.train</span>, ellipse <span class="op">=</span> <span class="cn">TRUE</span>,comp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">3</span><span class="op">)</span>,legend <span class="op">=</span> <span class="cn">TRUE</span>, title <span class="op">=</span><span class="st">"sample plot by mint"</span>, cex.main <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></code></pre></div>
<p><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/unnamed-chunk-38-1.png" width="700"></p>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#predict the rest of seqFISH</span>
<span class="va">mint.predict</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span><span class="op">(</span><span class="va">mint.train</span>, <span class="va">seqFISH_data_tobe_predicted</span>,
                        study.test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">seqFISH_data_tobe_predicted</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
<span class="va">seqFISH_the_rest</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">seqFISH_data_tobe_predicted</span>, <span class="va">mint.predict</span><span class="op">$</span><span class="va">class</span><span class="op">$</span><span class="va">max.dist</span><span class="op">[</span>, <span class="fl">6</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">seqFISH_the_rest</span><span class="op">)</span><span class="op">[</span><span class="fl">114</span><span class="op">]</span> <span class="op">=</span> <span class="st">"type"</span>
<span class="va">seqFISH_predicted</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">seqFISH_data_predicted</span>,<span class="va">seqFISH_label_predicted</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">seqFISH_predicted</span><span class="op">)</span><span class="op">[</span><span class="fl">114</span><span class="op">]</span> <span class="op">=</span> <span class="st">"type"</span>
<span class="va">final_prediction</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">seqFISH_the_rest</span>, <span class="va">seqFISH_predicted</span><span class="op">)</span>
<span class="va">final_prediction</span> <span class="op">=</span> <span class="va">final_prediction</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">final_prediction</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span>
<span class="va">final_labels</span> <span class="op">=</span> <span class="va">final_prediction</span><span class="op">$</span><span class="va">type</span></code></pre></div>
<p>The following figure shows the mean distribution of each cell type. The predicted cell types show distinct distribution patterns. Note that the mean distribution of GABA cells do not have DE genes. This is aligned with the observation that the predictive genes for GABA do not have very high loading coefficients.</p>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#plot the mean distribution of each class </span>
<span class="va">distribution</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">n_genes</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">final_prediction</span><span class="op">$</span><span class="va">type</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
  <span class="va">mean</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colMeans</a></span><span class="op">(</span><span class="va">final_prediction</span><span class="op">[</span><span class="va">final_prediction</span><span class="op">$</span><span class="va">type</span><span class="op">==</span><span class="va">i</span>,<span class="fl">1</span><span class="op">:</span><span class="va">n_genes</span><span class="op">]</span><span class="op">)</span>
  <span class="va">distribution</span>  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">distribution</span> ,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">mean</span>,<span class="va">i</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span>
<span class="va">distribution</span> <span class="op">=</span> <span class="va">distribution</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span>,<span class="op">]</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">n_comp</span><span class="op">)</span><span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">distribution</span><span class="op">[</span><span class="va">i</span>,<span class="op">-</span><span class="fl">1</span><span class="op">]</span>,ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">30</span>,<span class="fl">70</span><span class="op">)</span>, main <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"gene distribution of"</span>, 
                                                       <span class="va">distribution</span><span class="op">[</span><span class="va">i</span>, <span class="va">n_genes</span><span class="op">+</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>, 
       xlab <span class="op">=</span> <span class="st">"gene_id"</span>, ylab <span class="op">=</span> <span class="st">"normalised gene count"</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/unnamed-chunk-39-1.png" width="700"><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/unnamed-chunk-39-2.png" width="700"><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/unnamed-chunk-39-3.png" width="700"><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/unnamed-chunk-39-4.png" width="700"><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/unnamed-chunk-39-5.png" width="700"><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/unnamed-chunk-39-6.png" width="700"></p>
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seqFISH_relabel</span> <span class="op">=</span> <span class="va">seqFISH_label</span>
<span class="va">seqFISH_relabel</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">seqFISH_relabel</span> <span class="op">==</span> <span class="st">"Oligodendrocyte.1"</span><span class="op">)</span>, 
                  <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">seqFISH_relabel</span> <span class="op">==</span> <span class="st">"Oligodendrocyte.2"</span><span class="op">)</span>, 
                <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">seqFISH_relabel</span> <span class="op">==</span> <span class="st">"Oligodendrocyte.3"</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="op">=</span> <span class="st">"Oligodendrocyte"</span></code></pre></div>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">seqFISH_relabel</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">final_labels</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">final_labels</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 0.6537257</code></pre>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">converge_mint</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">seqFISH_relabel</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">final_labels</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">final_labels</span><span class="op">)</span></code></pre></div>
<p>65% of the predictions converge with Zhu et al. prediction. It is lower than the prediction made by pure sPLS-DA model. Note that Zhu’s prediction is not benchmarked. Here is not used for evaluation. For studies with no certain answers, we should compare results from different methods.</p>
</div>
<div id="conclusions" class="section level1">
<h1 class="hasAnchor">
<a href="#conclusions" class="anchor"></a>Conclusions</h1>
<p>The provided datasets have been well preprocessed. We further did quantile normalisation across all samples to try to eliminate batch effect before the supervised stage of learning. We also oversampled the data since the classes were imbalanced. The learning algorithm was divided into two analysis stages. First was to learn a classifier on scRNA data using sPLS-DA model and predict seqFISH data using the tuned <code>sPLS-DA</code> model. The second stage was to select cells with high probability of prediction as the new training data and combine them with scRNA data. The combined data were used to train a <code>mint.splsda</code> model to predict the rest of seqFISH data. Both models were cross validated (M-fold).</p>
<p>The selection of features were done by restricting the number of genes selected on each component. The final sPLS-Da model selected 24 genes to predict the cell types of the cells from the scRNA-seq data. The genes that contribute most to each classes were summarised and many of them were already referenced in the literature. The predictive top genes selected by MINT algorithms were in agreement with the top genes selected with sPLS-DA. It means that the batch effect does not influence much of the model training. The following table summarises the final prediction of seqFISH data. Finally, we visualised the mean gene distribution of the samples in each class. Each of the class presents a very distinct distribution pattern.</p>
<p>Around 71% of the sPLS-DA predictions converged with Zhu’s predictions. The semi-supervised approach by MINT had a lower convergence with Zhu’s paper (65%). However, these values are based on the assumption that Zhu’s prediction are considered as ‘ground truth’.</p>
<p>Here is the final prediction.</p>
<div class="sourceCode" id="cb101"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">final_labels</span><span class="op">)</span></code></pre></div>
<pre><code>## final_labels
##            Astrocyte     Endothelial Cell    GABA-ergic Neuron 
##                  176                   30                  449 
## Glutamatergic Neuron            Microglia      Oligodendrocyte 
##                  770                  103                   69</code></pre>

</div>
<div id="reference" class="section level1">
<h1 class="hasAnchor">
<a href="#reference" class="anchor"></a>Reference</h1>
<p>Barmeyer, C., Erko, I., Awad, K., Fromm, A., Bojarski, C., Meissner, S., Loddenkemper, C., Kerick, M., Siegmund, B., Fromm, M., Schweiger, M. R., &amp; Schulzke, J. D. (2017). Epithelial barrier dysfunction in lymphocytic colitis through cytokine-dependent internalization of claudin-5 and -8. Journal of gastroenterology, 52(10), 1090–1100. <a href="https://doi.org/10.1007/s00535-017-1309-2" class="uri">https://doi.org/10.1007/s00535-017-1309-2</a></p>
<p>Bonham, L. W., Sirkis, D. W., &amp; Yokoyama, J. S. (2019). The Transcriptional Landscape of Microglial Genes in Aging and Neurodegenerative Disease. Frontiers in immunology, 10, 1170. <a href="https://doi.org/10.3389/fimmu.2019.01170" class="uri">https://doi.org/10.3389/fimmu.2019.01170</a></p>
<p>Lê Cao, K. A., Boitard, S., &amp; Besse, P. (2011). Sparse PLS discriminant analysis: biologically relevant feature selection and graphical displays for multiclass problems. BMC bioinformatics, 12, 253. <a href="https://doi.org/10.1186/1471-2105-12-253" class="uri">https://doi.org/10.1186/1471-2105-12-253</a></p>
<p>Hevner, R. F., Shi, L., Justice, N., Hsueh, Y., Sheng, M., Smiga, S., Bulfone, A., Goffinet, A. M., Campagnoni, A. T., &amp; Rubenstein, J. L. (2001). Tbr1 regulates differentiation of the preplate and layer 6. Neuron, 29(2), 353–366. <a href="https://doi.org/10.1016/s0896-6273(01)00211-2" class="uri">https://doi.org/10.1016/s0896-6273(01)00211-2</a></p>
<p>Homkajorn, B., Sims, N. R., &amp; Muyderman, H. (2010). Connexin 43 regulates astrocytic migration and proliferation in response to injury. Neuroscience letters, 486(3), 197–201. <a href="https://doi.org/10.1016/j.neulet.2010.09.051" class="uri">https://doi.org/10.1016/j.neulet.2010.09.051</a></p>
<p>Rohart F., Gautier B., Singh A., and Le Cao K-A. (2017). mixOmics: An R package for ’omics feature selection and multiple data integration. PLoS computational biology 13(11):e1005752</p>
<p>Tasic, B., Menon, V., Nguyen, T. N., Kim, T. K., Jarsky, T., Yao, Z., Levi, B., Gray, L. T., Sorensen, S. A., Dolbeare, T., Bertagnolli, D., Goldy, J., Shapovalova, N., Parry, S., Lee, C., Smith, K., Bernard, A., Madisen, L., Sunkin, S. M., Hawrylycz, M., … Zeng, H. (2016). Adult mouse cortical cell taxonomy revealed by single cell transcriptomics. Nature neuroscience, 19(2), 335–346. <a href="https://doi.org/10.1038/nn.4216" class="uri">https://doi.org/10.1038/nn.4216</a></p>
<p>Zhu, Q., Shah, S., Dries, R., Cai, L., &amp; Yuan, G. C. (2018). Identification of spatially associated subpopulations by combining scRNAseq and sequential fluorescence in situ hybridization data. Nature biotechnology, 10.1038/nbt.4260. Advance online publication.</p>
<div style="page-break-after: always;"></div>
<div class="sourceCode" id="cb103"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R version 4.0.5 (2021-03-31)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.2 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0
## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0
## 
## locale:
##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] dplyr_1.0.6           preprocessCore_1.52.1 mixOmics_6.14.1      
## [4] ggplot2_3.3.3         lattice_0.20-41       MASS_7.3-53.1        
## [7] matrixStats_0.58.0   
## 
## loaded via a namespace (and not attached):
##  [1] ggrepel_0.9.1       Rcpp_1.0.6          tidyr_1.1.3        
##  [4] corpcor_1.6.9       rprojroot_2.0.2     digest_0.6.27      
##  [7] utf8_1.2.1          RSpectra_0.16-0     R6_2.5.0           
## [10] plyr_1.8.6          ellipse_0.4.2       evaluate_0.14      
## [13] highr_0.9           pillar_1.6.0        rlang_0.4.11       
## [16] Matrix_1.3-2        rmarkdown_2.8       pkgdown_1.6.1      
## [19] labeling_0.4.2      textshaping_0.3.3   desc_1.3.0         
## [22] rARPACK_0.11-0      BiocParallel_1.24.1 stringr_1.4.0      
## [25] igraph_1.2.6        munsell_0.5.0       compiler_4.0.5     
## [28] xfun_0.22           pkgconfig_2.0.3     systemfonts_1.0.1  
## [31] htmltools_0.5.1.1   tidyselect_1.1.1    tibble_3.1.1       
## [34] gridExtra_2.3       fansi_0.4.2         crayon_1.4.1       
## [37] withr_2.4.2         grid_4.0.5          gtable_0.3.0       
## [40] lifecycle_1.0.0     magrittr_2.0.1      scales_1.1.1       
## [43] stringi_1.5.3       cachem_1.0.4        farver_2.1.0       
## [46] reshape2_1.4.4      fs_1.5.0            ellipsis_0.3.2     
## [49] ragg_1.1.2          generics_0.1.0      vctrs_0.3.8        
## [52] RColorBrewer_1.1-2  tools_4.0.5         glue_1.4.2         
## [55] purrr_0.3.4         parallel_4.0.5      fastmap_1.1.0      
## [58] yaml_2.2.1          colorspace_2.0-1    memoise_2.0.0      
## [61] knitr_1.33</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Yuzhou Feng.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
