<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>BIRSBIO2020.scSpatial.sPLS-DA+MINT • BIRSBIO2020.scSpatial.sPLSDAMINT</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="BIRSBIO2020.scSpatial.sPLS-DA+MINT">
<meta property="og:description" content="BIRSBIO2020.scSpatial.sPLSDAMINT">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">BIRSBIO2020.scSpatial.sPLSDAMINT</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.9.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/BIRSBIO2020.scSpatial.sPLS-DA_MINT.html">BIRSBIO2020.scSpatial.sPLS-DA+MINT</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/fuerzhou/BIRSBIO2020.scSpatial.sPLS-DA-MINT">
    <span class="fas fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>BIRSBIO2020.scSpatial.sPLS-DA+MINT</h1>
                        <h4 class="author">Yuzhou Feng</h4>
            
            <h4 class="date">24/11/2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/fuerzhou/BIRSBIO2020.scSpatial.sPLS-DA-MINT/blob/master/vignettes/BIRSBIO2020.scSpatial.sPLS-DA_MINT.Rmd"><code>vignettes/BIRSBIO2020.scSpatial.sPLS-DA_MINT.Rmd</code></a></small>
      <div class="hidden name"><code>BIRSBIO2020.scSpatial.sPLS-DA_MINT.Rmd</code></div>

    </div>

    
    

<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<div id="background" class="section level2">
<h2 class="hasAnchor">
<a href="#background" class="anchor"></a>Background</h2>
<p>Tasic et al. (2016) used scRNA sequencing to explore the extent of cell types of mouse primal visual cortex. In the study, 1723 high-quality cells with thousands of genes in primal visual cortex in male adult mice were sequenced on a single cell resolution for cell classification. This study identified 49 cell types, including 23 GABAergic, 19 glutamatergic and 7 non-neural types in the 1723 cells.</p>
<p>The cell classification result from Tasic et al. (2016) was used by Zhu et al. (2018) to design a study to distinguish the difference between intrinsic and extrinsic effect on gene expression. Intrinsic effect means the regulatory gene network, while extrinsic means the cellular microenvironment. The study was conducted by combining scRNA sequencing data and smFISH data (single-molecule fluorescence in situ hybridization). The former one has high molecular resolution of transcriptomics (thousands of genes) without spatial information, while the latter keeps the spatial information but loses the high resolution (only a few hundred genes).</p>
<p>Zhu et al. (2018) mapped the sRNA sequencing data to the seqFISH data to enhance the molecular resolution of the cells with spatial information using SVM (support vector machine). The model was trained to identify the major cell type difference by training on scRNA data of 8 groups, GABAergic and Glutamatergic are the major neuron types, and other non-neuronal types, including Astrocytes, Mndothelial cells, microcytes, and three types of Oligocytes. The selected features (genes) were the top 43 differentially expressed (DE) genes in the identified 113 genes. The classification result was FURTHER validated by different evidence like cell type specific staining and previously reported marker genes.</p>
<p>Now, given a dataset of gene expression levels of 1723 cells and 113 genes and the cell types (from Tasic et al. 2016), our task is to classify another set of cells provided by Zhu et al. (2018), 1597 cells with 113 genes.</p>
<p><strong>Outline:</strong></p>
<p>I will adopt a semi-supervised approach to classify the cells of seqFISH data.</p>
<ol style="list-style-type: decimal">
<li>Train an sPLS-DA model on scRNA dataset.</li>
<li>Select genes by limiting the number of genes of “keepX” argument of each component during hyperparameter tuning.</li>
<li>Predict seqFISH data using the trained sPLS-DA model.</li>
<li>The predictions with high probabilities by the sPLS-DA model will be combined to the original training data for further training.</li>
<li>Train a MINT model on the combined training dataset and predict the rest cells of seqFISH data.</li>
<li>Identify the marker genes by examining the loading factors.</li>
</ol>
<p>This semi-supervised method can borrow the information from the seqFISH data and make the model customised. The combination of the data will be used to train a MINT.sPLS-DA model. MINT (Multivariate INTegrative method) (Rohart et al. 2017) is robust for integrating data from different sources regardless of the batch effect. The top discriminative genes identified from the model will be validated using previous literature as evidence. The determination of the minimal number of genes will be done by restricting the values of keepX and the performance will be monitored by balanced error rate (BER).</p>
</div>
<div id="data" class="section level2">
<h2 class="hasAnchor">
<a href="#data" class="anchor"></a>Data</h2>
<p>First import the data, including the training data (<code>scRNA</code>), the training labels (<code>scRNA_label</code>), the test data (<code>seqFISH</code>) and the predicted test labels by Zhu’s paper (<code>seqFISH_label</code>).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">##### import the data #####</span>
<span class="va">scRNA1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="st">"data/tasic_training_b2.txt"</span>,header <span class="op">=</span> <span class="cn">FALSE</span>, sep <span class="op">=</span><span class="st">'\t'</span><span class="op">)</span> 
<span class="va">scRNA</span> <span class="op">&lt;-</span> <span class="va">scRNA1</span><span class="op">[</span>,<span class="op">-</span><span class="fl">1</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">scRNA</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">scRNA1</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>
<span class="va">scRNA</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">scRNA</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">scRNA</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"scRNA"</span>,<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">scRNA</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>,<span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">scRNA</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span>
  
<span class="va">seqFISH1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="st">"data/seqfish_cortex_b2_testing.txt"</span>,header <span class="op">=</span> <span class="cn">FALSE</span>, sep <span class="op">=</span><span class="st">'\t'</span><span class="op">)</span> 
<span class="va">seqFISH</span> <span class="op">&lt;-</span> <span class="va">seqFISH1</span><span class="op">[</span>,<span class="op">-</span><span class="fl">1</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">seqFISH</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">seqFISH1</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>
<span class="va">seqFISH</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">seqFISH</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">seqFISH</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"seqFISH"</span>,<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">seqFISH</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>,<span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">seqFISH</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span>

<span class="va">scRNA_label</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="st">"data/tasic_labels.tsv"</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>
<span class="va">seqFISH_label</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="st">"data/seqfish_labels.tsv"</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span></code></pre></div>
<p><code>scRNA</code> is from Taisc et el. 2016. It contains the gene expression level of 1723 cells with 113 genes.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">scRNA</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 1723  113</code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">scRNA</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>##         abca15 abca9 acta2 adcy4 aldh3b2 amigo2
## scRNA_1     11    22    15    12      27     23
## scRNA_2     42    46    47    45      49     43
## scRNA_3     17    22    15    12      27    101
## scRNA_4     42    46    42    45      49     43
## scRNA_5     35    39    34    38      42     72
## scRNA_6     35    39    42    38      42     43</code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">n_genes</span> <span class="op">&lt;-</span> <span class="fl">113</span>
<span class="va">n_scRNA</span> <span class="op">&lt;-</span> <span class="fl">1723</span></code></pre></div>
<p><code>seqFISH</code> is from Zhu et el. 2018. It contains the gene expression level of 1597 cells with the 113 genes shared with <code>scRNA</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">seqFISH</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 1597  113</code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">seqFISH</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>##           abca15 abca9 acta2 adcy4 aldh3b2 amigo2
## seqFISH_1     68    41    25    39     101     93
## seqFISH_2     49    42    23    54      47     64
## seqFISH_3     50    38    16    37      41     93
## seqFISH_4     39    36    21    18      52     93
## seqFISH_5     31    47    29    37     101     93
## seqFISH_6     59    18    36    26      50     93</code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">n_seqFISH</span> <span class="op">&lt;-</span> <span class="fl">1597</span></code></pre></div>
<p><code>scRNA_label</code> has the class of the cells corresponding to <code>scRNA</code> data. The phenotyping was done by Tasic et al. and the labels are treated as ground truth in our analysis and are used for training models.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">scRNA_label</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 1723</code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">scRNA_label</span><span class="op">)</span></code></pre></div>
<pre><code>## scRNA_label
##            Astrocyte     Endothelial Cell    GABA-ergic Neuron 
##                   43                   29                  761 
## Glutamatergic Neuron            Microglia    Oligodendrocyte.1 
##                  812                   22                   19 
##    Oligodendrocyte.2    Oligodendrocyte.3 
##                    6                   31</code></pre>
<p>There are 8 classes in <code>scRNA_label</code>. The numbers of cells in each type are not balanced.</p>
<p><code>seqFISH_label</code> has the classes predicted for <code>seqFISH</code> data by Zhu et al. The predicitons have not been benchmarked, thus <code>seqFISH_label</code> will not be used in our analysis.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">seqFISH_label</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 1597</code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">seqFISH_label</span><span class="op">)</span></code></pre></div>
<pre><code>## seqFISH_label
##            Astrocyte     Endothelial Cell    GABA-ergic Neuron 
##                   97                   11                  556 
## Glutamatergic Neuron            Microglia    Oligodendrocyte.1 
##                  859                   22                    8 
##    Oligodendrocyte.2    Oligodendrocyte.3 
##                   21                   23</code></pre>
</div>
</div>
<div id="data-preprocessing" class="section level1">
<h1 class="hasAnchor">
<a href="#data-preprocessing" class="anchor"></a>Data preprocessing</h1>
<div id="quantile-normalisation" class="section level2">
<h2 class="hasAnchor">
<a href="#quantile-normalisation" class="anchor"></a>Quantile normalisation</h2>
<p>First, we check the gene expression distribution of each gene in <code>scRNA</code> and <code>seqFISH</code>.</p>
<p><img src="hist_sc1.png"><img src="hist_sc2.png"></p>
<p><img src="hist_fish1.png"><img src="hist_fish2.png"></p>
<p>The histograms show that the features (genes) are of the same scale and the distributions are almost normal. It means that there’s no need for further feature normalisation.</p>
<p>Now check the gene expression distribution of the 40 random samples in each dataset.</p>
<p><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/unnamed-chunk-9-1.png" width="768"><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/unnamed-chunk-9-2.png" width="768"></p>
<p>Figure 1. Gene expression distribution of each sample in scRNA data and seqFISH data.</p>
<p>The boxplots show that the the distributions of scRNA data have some sample-wise variation, while seqFISH samples have more consistent distributions.</p>
<p>Some other analyses in the hackalthon did quantile normalisation for paired genes to try to eliminate the differences across datasets. However, I insist that we should do <strong>quantile normalisation</strong> on <strong>samples</strong>. The rationale behind sample quantile normalisation is that although different samples have different differentially expressed (DE) genes, the majority of the genes should express at the same level. That means whichever genes are DE, the mean and quantile of each sample should be of the similar level.</p>
<p>Therefore, I will unify the quantile normalisation of samples between the datasets to maximumly eliminate the differences between the datasets. The normalisation was done by <code>normalize.quantiles</code> function from <code>preprocessCore</code> package.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## QUANTILE normalisation on all the samples</span>
<span class="va">all_data</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">scRNA</span>,<span class="va">seqFISH</span><span class="op">)</span>
<span class="va">qt_all_data</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu">normalize.quantiles</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">all_data</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">qt_all_data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">scRNA</span><span class="op">)</span>
<span class="co"># get the quantile normalised scRNA and seqFISH data</span>
<span class="va">qt_scRNA</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">qt_all_data</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">scRNA</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,<span class="op">]</span><span class="op">)</span>
<span class="va">qt_seqFISH</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">qt_all_data</span><span class="op">[</span><span class="op">(</span><span class="fl">1</span><span class="op">+</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">scRNA</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">scRNA</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">+</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">seqFISH</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span></code></pre></div>
<p>Then check 40 random samples from each dataset again.</p>
<p><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/unnamed-chunk-11-1.png" width="768"><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/unnamed-chunk-11-2.png" width="768"></p>
<p>Figure 2. Gene expression distribution of 40 random samples in scRNA data and seqFISH data after quantile normalisation.</p>
<p><code>qt_scRNA</code> is quantile normalised scRNA data and <code>qt_seqFISH</code> is quantile normalised seqFISH data.</p>
</div>
<div id="oversampling" class="section level2">
<h2 class="hasAnchor">
<a href="#oversampling" class="anchor"></a>Oversampling</h2>
<p>Let’s check the number of samples in each class.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">scRNA_label</span><span class="op">)</span></code></pre></div>
<pre><code>## scRNA_label
##            Astrocyte     Endothelial Cell    GABA-ergic Neuron 
##                   43                   29                  761 
## Glutamatergic Neuron            Microglia    Oligodendrocyte.1 
##                  812                   22                   19 
##    Oligodendrocyte.2    Oligodendrocyte.3 
##                    6                   31</code></pre>
<p>The cells of each type are heavily imbalanced in the training data. There is not enough training data for the minor group and the major groups will dominate the classification. Then the model will not have enough power to classify minor groups. To solve the problem, I used two strategies. First, I combined the three subgroups of Oligodendroctyes into one group. This is because in the Tasic paper, some cells were not identified as “core” cells for any Oligo but identified as “intermediate” cells and these three types were identified as close to each other. Also, Oligo.2 has too few samples. The second strategy is to oversample the minor groups.</p>
<p>First combine the three Oligo subgroups.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scRNA_relabel</span> <span class="op">=</span> <span class="va">scRNA_label</span>
<span class="va">scRNA_relabel</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">scRNA_relabel</span> <span class="op">==</span> <span class="st">"Oligodendrocyte.1"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">scRNA_relabel</span> <span class="op">==</span> <span class="st">"Oligodendrocyte.2"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">scRNA_relabel</span> <span class="op">==</span> <span class="st">"Oligodendrocyte.3"</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="op">=</span> <span class="st">"Oligodendrocyte"</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">scRNA_relabel</span><span class="op">)</span></code></pre></div>
<pre><code>## scRNA_relabel
##            Astrocyte     Endothelial Cell    GABA-ergic Neuron 
##                   43                   29                  761 
## Glutamatergic Neuron            Microglia      Oligodendrocyte 
##                  812                   22                   56</code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">qt_scRNA_Astro</span> <span class="op">=</span> <span class="va">qt_scRNA</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">scRNA_label</span> <span class="op">==</span> <span class="st">"Astrocyte"</span><span class="op">)</span>,<span class="op">]</span>
<span class="va">qt_scRNA_Endo</span> <span class="op">=</span> <span class="va">qt_scRNA</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">scRNA_label</span> <span class="op">==</span> <span class="st">"Endothelial Cell"</span><span class="op">)</span>,<span class="op">]</span>
<span class="va">qt_scRNA_GABA</span> <span class="op">=</span> <span class="va">qt_scRNA</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">scRNA_label</span> <span class="op">==</span> <span class="st">"GABA-ergic Neuron"</span><span class="op">)</span>,<span class="op">]</span>
<span class="va">qt_scRNA_Glu</span> <span class="op">=</span> <span class="va">qt_scRNA</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">scRNA_label</span> <span class="op">==</span> <span class="st">"Glutamatergic Neuron"</span><span class="op">)</span>,<span class="op">]</span>
<span class="va">qt_scRNA_Micro</span> <span class="op">=</span> <span class="va">qt_scRNA</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">scRNA_label</span> <span class="op">==</span> <span class="st">"Microglia"</span><span class="op">)</span>,<span class="op">]</span>
<span class="va">qt_scRNA_O1</span> <span class="op">=</span> <span class="va">qt_scRNA</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">scRNA_label</span> <span class="op">==</span> <span class="st">"Oligodendrocyte.1"</span><span class="op">)</span>,<span class="op">]</span>
<span class="va">qt_scRNA_O2</span> <span class="op">=</span> <span class="va">qt_scRNA</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">scRNA_label</span> <span class="op">==</span> <span class="st">"Oligodendrocyte.2"</span><span class="op">)</span>,<span class="op">]</span>
<span class="va">qt_scRNA_O3</span> <span class="op">=</span> <span class="va">qt_scRNA</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">scRNA_label</span> <span class="op">==</span> <span class="st">"Oligodendrocyte.3"</span><span class="op">)</span>,<span class="op">]</span>
<span class="va">qt_scRNA_Oligo</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">qt_scRNA_O1</span>, <span class="va">qt_scRNA_O2</span>,<span class="va">qt_scRNA_O3</span><span class="op">)</span></code></pre></div>
<p>Now, oversample the minor groups.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">com_qt_scRNA_label</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"Astrocyte"</span>, <span class="fl">43</span><span class="op">*</span><span class="fl">15</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"Endothelial Cell"</span>,<span class="fl">29</span><span class="op">*</span><span class="fl">25</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"GABA-ergic Neuron"</span>,<span class="fl">761</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"Glutamatergic Neuron"</span>, <span class="fl">812</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"Microglia"</span>,<span class="fl">22</span><span class="op">*</span><span class="fl">30</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"Oligodendrocyte"</span>, <span class="fl">56</span><span class="op">*</span><span class="fl">12</span><span class="op">)</span><span class="op">)</span>
<span class="va">com_qt_scRNA_label</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">com_qt_scRNA_label</span><span class="op">)</span></code></pre></div>
<p>We end up with around 700 samples in each class. The total number of cells after oversampling is 4275.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">com_qt_scRNA_label</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 4275</code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">com_qt_scRNA_label</span><span class="op">)</span></code></pre></div>
<pre><code>## com_qt_scRNA_label
##            Astrocyte     Endothelial Cell    GABA-ergic Neuron 
##                  645                  725                  761 
## Glutamatergic Neuron            Microglia      Oligodendrocyte 
##                  812                  660                  672</code></pre>
<p>Correspondingly, the <code>scRNA</code> data should also be resized. <code>com_qt_scRNA</code> is the oversampled scRNA data. It has 4275 samples.<code>com_qt_scRNA_label</code> is the new scRNA label vector.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">com_qt_scRNA_Astro</span> <span class="op">=</span> <span class="va">qt_scRNA_Astro</span> <span class="op">%&gt;%</span> <span class="fu">slice</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu">n</span><span class="op">(</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">15</span><span class="op">)</span><span class="op">)</span>
<span class="va">com_qt_scRNA_Endo</span> <span class="op">=</span> <span class="va">qt_scRNA_Endo</span> <span class="op">%&gt;%</span> <span class="fu">slice</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu">n</span><span class="op">(</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">25</span><span class="op">)</span><span class="op">)</span>
<span class="va">com_qt_scRNA_GABA</span> <span class="op">=</span> <span class="va">qt_scRNA_GABA</span> 
<span class="va">com_qt_scRNA_Glu</span> <span class="op">=</span> <span class="va">qt_scRNA_Glu</span>
<span class="va">com_qt_scRNA_Micro</span> <span class="op">=</span> <span class="va">qt_scRNA_Micro</span> <span class="op">%&gt;%</span> <span class="fu">slice</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu">n</span><span class="op">(</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">30</span><span class="op">)</span><span class="op">)</span>
<span class="va">com_qt_scRNA_Oligo</span> <span class="op">=</span> <span class="va">qt_scRNA_Oligo</span> <span class="op">%&gt;%</span> <span class="fu">slice</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu">n</span><span class="op">(</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">12</span><span class="op">)</span><span class="op">)</span>
<span class="va">com_qt_scRNA</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">com_qt_scRNA_Astro</span>, <span class="va">com_qt_scRNA_Endo</span>, <span class="va">com_qt_scRNA_GABA</span>, <span class="va">com_qt_scRNA_Glu</span>,
                     <span class="va">com_qt_scRNA_Micro</span>, <span class="va">com_qt_scRNA_Oligo</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">com_qt_scRNA</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 4275  113</code></pre>
</div>
</div>
<div id="train-an-spls-da-model" class="section level1">
<h1 class="hasAnchor">
<a href="#train-an-spls-da-model" class="anchor"></a>Train an sPLS-DA model</h1>
<p>Sparse PLS discriminant analysis (sPLS-DA) is a supervised machine learning algorithm (Lê Cao et al. 2011). Its performance in public microarray and SNP data sets is similar to other algorithms and it’s very efficient. The function is available under <code>mixOmics</code> package.</p>
<p>Train the sPLS-DA model on preprocessed scRNA data. First tune the number of components.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#load('com_qt_tune-SPLSDA.RData')</span>
<span class="co"># build the model</span>
<span class="va">com_qt_splsda</span> <span class="op">=</span> <span class="fu">splsda</span><span class="op">(</span><span class="va">com_qt_scRNA</span>, <span class="va">com_qt_scRNA_label</span>, ncomp <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>
<span class="co"># hyperparameter tuning</span>
<span class="va">com_qt_perf.splsda</span> <span class="op">&lt;-</span> <span class="fu">perf</span><span class="op">(</span><span class="va">com_qt_splsda</span>, validation <span class="op">=</span> <span class="st">"Mfold"</span>, folds <span class="op">=</span> <span class="fl">5</span>, nrepeat <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">com_qt_perf.splsda</span>, overlay <span class="op">=</span> <span class="st">'measure'</span><span class="op">)</span></code></pre></div>
<p><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/unnamed-chunk-19-1.png" width="576"></p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">n_comp</span> <span class="op">&lt;-</span> <span class="fl">6</span>
<span class="va">error_rate</span> <span class="op">&lt;-</span> <span class="va">com_qt_perf.splsda</span><span class="op">$</span><span class="va">error.rate</span><span class="op">$</span><span class="va">BER</span><span class="op">[</span><span class="va">n_comp</span>,<span class="st">"max.dist"</span><span class="op">]</span></code></pre></div>
<p>The error rate curve for tuning the number of components shows that the elbow is at comp 6, and at comp 6 <strong>max distance</strong> has the lowest error rate 0.052495. It’s worth noting that after oversampling, the balanced error rate (BER) is similar to overall error rate. I still use BER since the numbers of each class are still different.</p>
<p>Then, tune the number of genes of each compnent. Use <strong>max distance</strong> as the distance method.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">list.keepX</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">100</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span>
<span class="va">com_qt_splsda.tune</span> <span class="op">&lt;-</span> <span class="fu">tune.splsda</span><span class="op">(</span><span class="va">com_qt_scRNA</span>, <span class="va">com_qt_scRNA_label</span>, ncomp <span class="op">=</span> <span class="fl">6</span>, validation <span class="op">=</span> <span class="st">'Mfold'</span>, folds <span class="op">=</span> <span class="fl">5</span>, dist <span class="op">=</span> <span class="st">'max.dist'</span>,progressBar <span class="op">=</span> <span class="cn">T</span>  , test.keepX <span class="op">=</span> <span class="va">list.keepX</span>, nrepeat <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></code></pre></div>
<pre><code>## 
## comp 1 
## 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |==                                                                    |   2%
  |                                                                            
  |====                                                                  |   5%
  |                                                                            
  |=====                                                                 |   8%
  |                                                                            
  |=======                                                               |  10%
  |                                                                            
  |=========                                                             |  12%
  |                                                                            
  |==========                                                            |  15%
  |                                                                            
  |============                                                          |  18%
  |                                                                            
  |==============                                                        |  20%
  |                                                                            
  |================                                                      |  22%
  |                                                                            
  |==================                                                    |  25%
  |                                                                            
  |===================                                                   |  28%
  |                                                                            
  |=====================                                                 |  30%
  |                                                                            
  |=======================                                               |  32%
  |                                                                            
  |========================                                              |  35%
  |                                                                            
  |==========================                                            |  38%
  |                                                                            
  |============================                                          |  40%
  |                                                                            
  |==============================                                        |  42%
  |                                                                            
  |================================                                      |  45%
  |                                                                            
  |=================================                                     |  48%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |=====================================                                 |  52%
  |                                                                            
  |======================================                                |  55%
  |                                                                            
  |========================================                              |  57%
  |                                                                            
  |==========================================                            |  60%
  |                                                                            
  |============================================                          |  62%
  |                                                                            
  |==============================================                        |  65%
  |                                                                            
  |===============================================                       |  68%
  |                                                                            
  |=================================================                     |  70%
  |                                                                            
  |===================================================                   |  72%
  |                                                                            
  |====================================================                  |  75%
  |                                                                            
  |======================================================                |  78%
  |                                                                            
  |========================================================              |  80%
  |                                                                            
  |==========================================================            |  82%
  |                                                                            
  |============================================================          |  85%
  |                                                                            
  |=============================================================         |  88%
  |                                                                            
  |===============================================================       |  90%
  |                                                                            
  |=================================================================     |  92%
  |                                                                            
  |==================================================================    |  95%
  |                                                                            
  |====================================================================  |  98%
  |                                                                            
  |======================================================================| 100%
## comp 2 
## 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |==                                                                    |   2%
  |                                                                            
  |====                                                                  |   5%
  |                                                                            
  |=====                                                                 |   8%
  |                                                                            
  |=======                                                               |  10%
  |                                                                            
  |=========                                                             |  12%
  |                                                                            
  |==========                                                            |  15%
  |                                                                            
  |============                                                          |  18%
  |                                                                            
  |==============                                                        |  20%
  |                                                                            
  |================                                                      |  22%
  |                                                                            
  |==================                                                    |  25%
  |                                                                            
  |===================                                                   |  28%
  |                                                                            
  |=====================                                                 |  30%
  |                                                                            
  |=======================                                               |  32%
  |                                                                            
  |========================                                              |  35%
  |                                                                            
  |==========================                                            |  38%
  |                                                                            
  |============================                                          |  40%
  |                                                                            
  |==============================                                        |  42%
  |                                                                            
  |================================                                      |  45%
  |                                                                            
  |=================================                                     |  48%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |=====================================                                 |  52%
  |                                                                            
  |======================================                                |  55%
  |                                                                            
  |========================================                              |  57%
  |                                                                            
  |==========================================                            |  60%
  |                                                                            
  |============================================                          |  62%
  |                                                                            
  |==============================================                        |  65%
  |                                                                            
  |===============================================                       |  68%
  |                                                                            
  |=================================================                     |  70%
  |                                                                            
  |===================================================                   |  72%
  |                                                                            
  |====================================================                  |  75%
  |                                                                            
  |======================================================                |  78%
  |                                                                            
  |========================================================              |  80%
  |                                                                            
  |==========================================================            |  82%
  |                                                                            
  |============================================================          |  85%
  |                                                                            
  |=============================================================         |  88%
  |                                                                            
  |===============================================================       |  90%
  |                                                                            
  |=================================================================     |  92%
  |                                                                            
  |==================================================================    |  95%
  |                                                                            
  |====================================================================  |  98%
  |                                                                            
  |======================================================================| 100%
## comp 3 
## 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |==                                                                    |   2%
  |                                                                            
  |====                                                                  |   5%
  |                                                                            
  |=====                                                                 |   8%
  |                                                                            
  |=======                                                               |  10%
  |                                                                            
  |=========                                                             |  12%
  |                                                                            
  |==========                                                            |  15%
  |                                                                            
  |============                                                          |  18%
  |                                                                            
  |==============                                                        |  20%
  |                                                                            
  |================                                                      |  22%
  |                                                                            
  |==================                                                    |  25%
  |                                                                            
  |===================                                                   |  28%
  |                                                                            
  |=====================                                                 |  30%
  |                                                                            
  |=======================                                               |  32%
  |                                                                            
  |========================                                              |  35%
  |                                                                            
  |==========================                                            |  38%
  |                                                                            
  |============================                                          |  40%
  |                                                                            
  |==============================                                        |  42%
  |                                                                            
  |================================                                      |  45%
  |                                                                            
  |=================================                                     |  48%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |=====================================                                 |  52%
  |                                                                            
  |======================================                                |  55%
  |                                                                            
  |========================================                              |  57%
  |                                                                            
  |==========================================                            |  60%
  |                                                                            
  |============================================                          |  62%
  |                                                                            
  |==============================================                        |  65%
  |                                                                            
  |===============================================                       |  68%
  |                                                                            
  |=================================================                     |  70%
  |                                                                            
  |===================================================                   |  72%
  |                                                                            
  |====================================================                  |  75%
  |                                                                            
  |======================================================                |  78%
  |                                                                            
  |========================================================              |  80%
  |                                                                            
  |==========================================================            |  82%
  |                                                                            
  |============================================================          |  85%
  |                                                                            
  |=============================================================         |  88%
  |                                                                            
  |===============================================================       |  90%
  |                                                                            
  |=================================================================     |  92%
  |                                                                            
  |==================================================================    |  95%
  |                                                                            
  |====================================================================  |  98%
  |                                                                            
  |======================================================================| 100%
## comp 4 
## 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |==                                                                    |   2%
  |                                                                            
  |====                                                                  |   5%
  |                                                                            
  |=====                                                                 |   8%
  |                                                                            
  |=======                                                               |  10%
  |                                                                            
  |=========                                                             |  12%
  |                                                                            
  |==========                                                            |  15%
  |                                                                            
  |============                                                          |  18%
  |                                                                            
  |==============                                                        |  20%
  |                                                                            
  |================                                                      |  22%
  |                                                                            
  |==================                                                    |  25%
  |                                                                            
  |===================                                                   |  28%
  |                                                                            
  |=====================                                                 |  30%
  |                                                                            
  |=======================                                               |  32%
  |                                                                            
  |========================                                              |  35%
  |                                                                            
  |==========================                                            |  38%
  |                                                                            
  |============================                                          |  40%
  |                                                                            
  |==============================                                        |  42%
  |                                                                            
  |================================                                      |  45%
  |                                                                            
  |=================================                                     |  48%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |=====================================                                 |  52%
  |                                                                            
  |======================================                                |  55%
  |                                                                            
  |========================================                              |  57%
  |                                                                            
  |==========================================                            |  60%
  |                                                                            
  |============================================                          |  62%
  |                                                                            
  |==============================================                        |  65%
  |                                                                            
  |===============================================                       |  68%
  |                                                                            
  |=================================================                     |  70%
  |                                                                            
  |===================================================                   |  72%
  |                                                                            
  |====================================================                  |  75%
  |                                                                            
  |======================================================                |  78%
  |                                                                            
  |========================================================              |  80%
  |                                                                            
  |==========================================================            |  82%
  |                                                                            
  |============================================================          |  85%
  |                                                                            
  |=============================================================         |  88%
  |                                                                            
  |===============================================================       |  90%
  |                                                                            
  |=================================================================     |  92%
  |                                                                            
  |==================================================================    |  95%
  |                                                                            
  |====================================================================  |  98%
  |                                                                            
  |======================================================================| 100%
## comp 5 
## 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |==                                                                    |   2%
  |                                                                            
  |====                                                                  |   5%
  |                                                                            
  |=====                                                                 |   8%
  |                                                                            
  |=======                                                               |  10%
  |                                                                            
  |=========                                                             |  12%
  |                                                                            
  |==========                                                            |  15%
  |                                                                            
  |============                                                          |  18%
  |                                                                            
  |==============                                                        |  20%
  |                                                                            
  |================                                                      |  22%
  |                                                                            
  |==================                                                    |  25%
  |                                                                            
  |===================                                                   |  28%
  |                                                                            
  |=====================                                                 |  30%
  |                                                                            
  |=======================                                               |  32%
  |                                                                            
  |========================                                              |  35%
  |                                                                            
  |==========================                                            |  38%
  |                                                                            
  |============================                                          |  40%
  |                                                                            
  |==============================                                        |  42%
  |                                                                            
  |================================                                      |  45%
  |                                                                            
  |=================================                                     |  48%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |=====================================                                 |  52%
  |                                                                            
  |======================================                                |  55%
  |                                                                            
  |========================================                              |  57%
  |                                                                            
  |==========================================                            |  60%
  |                                                                            
  |============================================                          |  62%
  |                                                                            
  |==============================================                        |  65%
  |                                                                            
  |===============================================                       |  68%
  |                                                                            
  |=================================================                     |  70%
  |                                                                            
  |===================================================                   |  72%
  |                                                                            
  |====================================================                  |  75%
  |                                                                            
  |======================================================                |  78%
  |                                                                            
  |========================================================              |  80%
  |                                                                            
  |==========================================================            |  82%
  |                                                                            
  |============================================================          |  85%
  |                                                                            
  |=============================================================         |  88%
  |                                                                            
  |===============================================================       |  90%
  |                                                                            
  |=================================================================     |  92%
  |                                                                            
  |==================================================================    |  95%
  |                                                                            
  |====================================================================  |  98%
  |                                                                            
  |======================================================================| 100%
## comp 6 
## 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |==                                                                    |   2%
  |                                                                            
  |====                                                                  |   5%
  |                                                                            
  |=====                                                                 |   8%
  |                                                                            
  |=======                                                               |  10%
  |                                                                            
  |=========                                                             |  12%
  |                                                                            
  |==========                                                            |  15%
  |                                                                            
  |============                                                          |  18%
  |                                                                            
  |==============                                                        |  20%
  |                                                                            
  |================                                                      |  22%
  |                                                                            
  |==================                                                    |  25%
  |                                                                            
  |===================                                                   |  28%
  |                                                                            
  |=====================                                                 |  30%
  |                                                                            
  |=======================                                               |  32%
  |                                                                            
  |========================                                              |  35%
  |                                                                            
  |==========================                                            |  38%
  |                                                                            
  |============================                                          |  40%
  |                                                                            
  |==============================                                        |  42%
  |                                                                            
  |================================                                      |  45%
  |                                                                            
  |=================================                                     |  48%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |=====================================                                 |  52%
  |                                                                            
  |======================================                                |  55%
  |                                                                            
  |========================================                              |  57%
  |                                                                            
  |==========================================                            |  60%
  |                                                                            
  |============================================                          |  62%
  |                                                                            
  |==============================================                        |  65%
  |                                                                            
  |===============================================                       |  68%
  |                                                                            
  |=================================================                     |  70%
  |                                                                            
  |===================================================                   |  72%
  |                                                                            
  |====================================================                  |  75%
  |                                                                            
  |======================================================                |  78%
  |                                                                            
  |========================================================              |  80%
  |                                                                            
  |==========================================================            |  82%
  |                                                                            
  |============================================================          |  85%
  |                                                                            
  |=============================================================         |  88%
  |                                                                            
  |===============================================================       |  90%
  |                                                                            
  |=================================================================     |  92%
  |                                                                            
  |==================================================================    |  95%
  |                                                                            
  |====================================================================  |  98%
  |                                                                            
  |======================================================================| 100%</code></pre>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">com_qt_splsda.tune</span>, optimal <span class="op">=</span> <span class="cn">TRUE</span>, sd <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="BIRSBIO2020.scSpatial.sPLS-DA_MINT_files/figure-html/unnamed-chunk-20-1.png" width="480"></p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#save(com_qt_splsda, com_qt_perf.splsda, com_qt_splsda.tune, file = 'com_qt_tune-SPLSDA.RData')</span>
<span class="va">com_qt_splsda.tune</span><span class="op">$</span><span class="va">choice.keepX</span></code></pre></div>
<pre><code>## comp1 comp2 comp3 comp4 comp5 comp6 
##     5     1    65     1    75    10</code></pre>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">keepX</span> <span class="op">&lt;-</span> <span class="va">com_qt_splsda.tune</span><span class="op">$</span><span class="va">choice.keepX</span>
<span class="va">com_qt_splsda.tune</span><span class="op">$</span><span class="va">error.rate</span><span class="op">[</span><span class="fl">27</span>,<span class="fl">6</span><span class="op">]</span></code></pre></div>
<pre><code>## [1] 0.1039497</code></pre>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">error_rate</span> <span class="op">&lt;-</span> <span class="va">com_qt_splsda.tune</span><span class="op">$</span><span class="va">error.rate</span><span class="op">[</span><span class="fl">27</span>,<span class="fl">6</span><span class="op">]</span></code></pre></div>
<p>The number of selected genes in each component is 5, 1, 65, 1, 75, 10. The BER (balanced error rate, max distance) is 0.1039497.</p>
<div style="page-break-after: always;"></div>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R version 4.0.3 (2020-10-10)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.1 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] dplyr_1.0.2           preprocessCore_1.52.0 mixOmics_6.14.0      
## [4] ggplot2_3.3.3         lattice_0.20-41       MASS_7.3-53          
## [7] matrixStats_0.57.0   
## 
## loaded via a namespace (and not attached):
##  [1] tidyselect_1.1.0   xfun_0.20          purrr_0.3.4        reshape2_1.4.4    
##  [5] colorspace_2.0-0   vctrs_0.3.6        generics_0.1.0     htmltools_0.5.0   
##  [9] yaml_2.2.1         rlang_0.4.10       pkgdown_1.6.1      pillar_1.4.7      
## [13] glue_1.4.2         withr_2.3.0        RColorBrewer_1.1-2 lifecycle_0.2.0   
## [17] plyr_1.8.6         stringr_1.4.0      munsell_0.5.0      gtable_0.3.0      
## [21] ragg_0.4.0         memoise_1.1.0      evaluate_0.14      labeling_0.4.2    
## [25] knitr_1.30         parallel_4.0.3     rARPACK_0.11-0     Rcpp_1.0.5        
## [29] corpcor_1.6.9      scales_1.1.1       desc_1.2.0         farver_2.0.3      
## [33] systemfonts_0.3.2  fs_1.5.0           RSpectra_0.16-0    textshaping_0.2.1 
## [37] gridExtra_2.3      ellipse_0.4.2      digest_0.6.27      stringi_1.5.3     
## [41] ggrepel_0.9.0      grid_4.0.3         rprojroot_2.0.2    tools_4.0.3       
## [45] magrittr_2.0.1     tibble_3.0.4       crayon_1.3.4       tidyr_1.1.2       
## [49] pkgconfig_2.0.3    ellipsis_0.3.1     Matrix_1.3-2       assertthat_0.2.1  
## [53] rmarkdown_2.6      R6_2.5.0           igraph_1.2.6       compiler_4.0.3</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Yuzhou Feng.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
